{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "machine_shape": "hm"
    },
    "kernelspec": {
      "name": "ir",
      "display_name": "R"
    },
    "language_info": {
      "name": "R"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "source": [
        "install.packages(\"rjags\")\n",
        "devtools::install_github(\"broadinstitute/infercnv\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "MQFxWUtrnsXk",
        "outputId": "4f4e610d-a544-45be-9ae3-fd3bd926f99c"
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n",
            "also installing the dependency ‘coda’\n",
            "\n",
            "\n",
            "Downloading GitHub repo broadinstitute/infercnv@HEAD\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "XVector      (NA -> 0.48.0   ) [CRAN]\n",
            "abind        (NA -> 1.4-8    ) [CRAN]\n",
            "matrixStats  (NA -> 1.5.0    ) [CRAN]\n",
            "SparseArray  (NA -> 1.8.0    ) [CRAN]\n",
            "S4Arrays     (NA -> 1.8.0    ) [CRAN]\n",
            "IRanges      (NA -> 2.42.0   ) [CRAN]\n",
            "S4Vectors    (NA -> 0.46.0   ) [CRAN]\n",
            "MatrixGen... (NA -> 1.20.0   ) [CRAN]\n",
            "BiocGenerics (NA -> 0.54.0   ) [CRAN]\n",
            "GenomeInf... (NA -> 1.2.14   ) [CRAN]\n",
            "UCSC.utils   (NA -> 1.4.0    ) [CRAN]\n",
            "GenomeInfoDb (NA -> 1.44.0   ) [CRAN]\n",
            "DelayedArray (NA -> 0.34.1   ) [CRAN]\n",
            "Biobase      (NA -> 2.68.0   ) [CRAN]\n",
            "GenomicRa... (NA -> 1.60.0   ) [CRAN]\n",
            "RcppEigen    (NA -> 0.3.4.0.2) [CRAN]\n",
            "sitmo        (NA -> 2.0.2    ) [CRAN]\n",
            "BH           (NA -> 1.87.0-1 ) [CRAN]\n",
            "RcppProgress (NA -> 0.4.2    ) [CRAN]\n",
            "dqrng        (NA -> 0.4.1    ) [CRAN]\n",
            "RSpectra     (NA -> 0.16-2   ) [CRAN]\n",
            "RcppAnnoy    (NA -> 0.0.22   ) [CRAN]\n",
            "irlba        (NA -> 2.3.5.1  ) [CRAN]\n",
            "FNN          (NA -> 1.1.4.1  ) [CRAN]\n",
            "polyclip     (NA -> 1.10-7   ) [CRAN]\n",
            "deldir       (NA -> 2.0-4    ) [CRAN]\n",
            "spatstat.... (NA -> 3.1-3    ) [CRAN]\n",
            "spatstat.... (NA -> 3.1-2    ) [CRAN]\n",
            "spatstat.... (NA -> 3.1-6    ) [CRAN]\n",
            "tensor       (NA -> 1.5      ) [CRAN]\n",
            "goftest      (NA -> 1.2-3    ) [CRAN]\n",
            "spatstat.... (NA -> 3.1-0    ) [CRAN]\n",
            "spatstat.... (NA -> 3.3-3    ) [CRAN]\n",
            "spatstat.... (NA -> 3.3-6    ) [CRAN]\n",
            "plyr         (NA -> 1.8.9    ) [CRAN]\n",
            "parallelly   (NA -> 1.44.0   ) [CRAN]\n",
            "listenv      (NA -> 0.9.1    ) [CRAN]\n",
            "globals      (NA -> 0.17.0   ) [CRAN]\n",
            "RcppArmad... (NA -> 14.4.2-1 ) [CRAN]\n",
            "gridExtra    (NA -> 2.3      ) [CRAN]\n",
            "reshape2     (NA -> 1.4.4    ) [CRAN]\n",
            "future       (NA -> 1.40.0   ) [CRAN]\n",
            "future.apply (NA -> 1.11.3   ) [CRAN]\n",
            "bitops       (NA -> 1.0-9    ) [CRAN]\n",
            "caTools      (NA -> 1.18.3   ) [CRAN]\n",
            "gtools       (NA -> 3.9.5    ) [CRAN]\n",
            "gplots       (NA -> 3.2.0    ) [CRAN]\n",
            "png          (NA -> 0.1-8    ) [CRAN]\n",
            "here         (NA -> 1.0.1    ) [CRAN]\n",
            "RcppTOML     (NA -> 0.2.3    ) [CRAN]\n",
            "crosstalk    (NA -> 1.2.1    ) [CRAN]\n",
            "lazyeval     (NA -> 0.2.2    ) [CRAN]\n",
            "zoo          (NA -> 1.8-14   ) [CRAN]\n",
            "igraph       (NA -> 2.1.4    ) [CRAN]\n",
            "dotCall64    (NA -> 1.2      ) [CRAN]\n",
            "spam         (NA -> 2.11-1   ) [CRAN]\n",
            "progressr    (NA -> 0.15.1   ) [CRAN]\n",
            "sp           (NA -> 2.2-0    ) [CRAN]\n",
            "iterators    (NA -> 1.0.14   ) [CRAN]\n",
            "sandwich     (NA -> 3.1-1    ) [CRAN]\n",
            "TH.data      (NA -> 1.1-3    ) [CRAN]\n",
            "mvtnorm      (NA -> 1.3-3    ) [CRAN]\n",
            "statmod      (NA -> 1.5.0    ) [CRAN]\n",
            "formatR      (NA -> 1.14     ) [CRAN]\n",
            "findpython   (NA -> 1.0.9    ) [CRAN]\n",
            "Summarize... (NA -> 1.38.1   ) [CRAN]\n",
            "uwot         (NA -> 0.2.3    ) [CRAN]\n",
            "spatstat.... (NA -> 3.4-2    ) [CRAN]\n",
            "sctransform  (NA -> 0.4.2    ) [CRAN]\n",
            "scattermore  (NA -> 1.2      ) [CRAN]\n",
            "Rtsne        (NA -> 0.17     ) [CRAN]\n",
            "ROCR         (NA -> 1.0-11   ) [CRAN]\n",
            "reticulate   (NA -> 1.42.0   ) [CRAN]\n",
            "RcppHNSW     (NA -> 0.6.0    ) [CRAN]\n",
            "RANN         (NA -> 2.6.2    ) [CRAN]\n",
            "plotly       (NA -> 4.10.4   ) [CRAN]\n",
            "pbapply      (NA -> 1.7-2    ) [CRAN]\n",
            "patchwork    (NA -> 1.3.0    ) [CRAN]\n",
            "lmtest       (NA -> 0.9-40   ) [CRAN]\n",
            "leidenbase   (NA -> 0.1.35   ) [CRAN]\n",
            "ica          (NA -> 1.0-3    ) [CRAN]\n",
            "ggridges     (NA -> 0.5.6    ) [CRAN]\n",
            "ggrepel      (NA -> 0.9.6    ) [CRAN]\n",
            "fitdistrplus (NA -> 1.2-2    ) [CRAN]\n",
            "fastDummies  (NA -> 1.7.5    ) [CRAN]\n",
            "cowplot      (NA -> 1.1.3    ) [CRAN]\n",
            "SeuratObject (NA -> 5.1.0    ) [CRAN]\n",
            "foreach      (NA -> 1.5.2    ) [CRAN]\n",
            "multcomp     (NA -> 1.4-28   ) [CRAN]\n",
            "modeltools   (NA -> 0.2-24   ) [CRAN]\n",
            "libcoin      (NA -> 1.0-10   ) [CRAN]\n",
            "locfit       (NA -> 1.5-9.12 ) [CRAN]\n",
            "limma        (NA -> 3.64.0   ) [CRAN]\n",
            "RcppParallel (NA -> 5.1.10   ) [CRAN]\n",
            "ape          (NA -> 5.8-1    ) [CRAN]\n",
            "futile.op... (NA -> 1.0.1    ) [CRAN]\n",
            "lambda.r     (NA -> 1.2.4    ) [CRAN]\n",
            "argparse     (NA -> 2.2.5    ) [CRAN]\n",
            "SingleCel... (NA -> 1.30.0   ) [CRAN]\n",
            "Seurat       (NA -> 5.3.0    ) [CRAN]\n",
            "doParallel   (NA -> 1.0.17   ) [CRAN]\n",
            "coin         (NA -> 1.4-3    ) [CRAN]\n",
            "edgeR        (NA -> 4.6.1    ) [CRAN]\n",
            "HiddenMarkov (NA -> 1.8-14   ) [CRAN]\n",
            "parallelDist (NA -> 0.2.6    ) [CRAN]\n",
            "fastcluster  (NA -> 1.3.0    ) [CRAN]\n",
            "phyclust     (NA -> 0.1-34   ) [CRAN]\n",
            "futile.lo... (NA -> 1.4.3    ) [CRAN]\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Installing 108 packages: XVector, abind, matrixStats, SparseArray, S4Arrays, IRanges, S4Vectors, MatrixGenerics, BiocGenerics, GenomeInfoDbData, UCSC.utils, GenomeInfoDb, DelayedArray, Biobase, GenomicRanges, RcppEigen, sitmo, BH, RcppProgress, dqrng, RSpectra, RcppAnnoy, irlba, FNN, polyclip, deldir, spatstat.utils, spatstat.univar, spatstat.data, tensor, goftest, spatstat.sparse, spatstat.random, spatstat.geom, plyr, parallelly, listenv, globals, RcppArmadillo, gridExtra, reshape2, future, future.apply, bitops, caTools, gtools, gplots, png, here, RcppTOML, crosstalk, lazyeval, zoo, igraph, dotCall64, spam, progressr, sp, iterators, sandwich, TH.data, mvtnorm, statmod, formatR, findpython, SummarizedExperiment, uwot, spatstat.explore, sctransform, scattermore, Rtsne, ROCR, reticulate, RcppHNSW, RANN, plotly, pbapply, patchwork, lmtest, leidenbase, ica, ggridges, ggrepel, fitdistrplus, fastDummies, cowplot, SeuratObject, foreach, multcomp, modeltools, libcoin, locfit, limma, RcppParallel, ape, futile.options, lambda.r, argparse, SingleCellExperiment, Seurat, doParallel, coin, edgeR, HiddenMarkov, parallelDist, fastcluster, phyclust, futile.logger\n",
            "\n",
            "Installing packages into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b[36m──\u001b[39m \u001b[36mR CMD build\u001b[39m \u001b[36m─────────────────────────────────────────────────────────────────\u001b[39m\n",
            "* checking for file ‘/tmp/RtmpCug5rq/remotes15f06a80bd3d/broadinstitute-infercnv-624feae/DESCRIPTION’ ... OK\n",
            "* preparing ‘infercnv’:\n",
            "* checking DESCRIPTION meta-information ... OK\n",
            "* checking for LF line-endings in source and make files and shell scripts\n",
            "* checking for empty or unneeded directories\n",
            "Removed empty directory ‘infercnv/WDL’\n",
            "* building ‘infercnv_1.23.0.tar.gz’\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "library(infercnv)"
      ],
      "metadata": {
        "id": "y72Zp_WNZLCB"
      },
      "execution_count": 2,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "infercnv_obj <- infercnv::CreateInfercnvObject(\n",
        "  raw_counts_matrix = \"expr_matrix.txt\",\n",
        "  gene_order_file = \"gene_order.txt\",\n",
        "  annotations_file =\"cell_annotations.txt\",\n",
        "  ref_group_names = c(\"B cell\", \"CD4 T cell\")\n",
        ")\n",
        "\n",
        "infercnv_obj = infercnv::run(infercnv_obj,\n",
        "                             cutoff=1, # cutoff=1 works well for Smart-seq2, and cutoff=0.1 works well for 10x Genomics\n",
        "                             out_dir= 'infercnv_output',\n",
        "                             cluster_by_groups=TRUE,\n",
        "                             denoise=TRUE,\n",
        "                             HMM=TRUE)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "8_WcmyMGs5AG",
        "outputId": "139de692-5a6c-4ee3-aaec-9f761020b476"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-05-07 17:30:36] Parsing matrix: expr_matrix.txt\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message in scan(file = file, what = what, sep = sep, quote = quote, dec = dec, :\n",
            "“number of items read is not a multiple of the number of columns”\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-05-07 17:31:06] Parsing gene order file: gene_order.txt\n",
            "INFO [2025-05-07 17:31:06] Parsing cell annotations file: cell_annotations.txt\n",
            "INFO [2025-05-07 17:31:06] ::order_reduce:Start.\n",
            "INFO [2025-05-07 17:31:06] .order_reduce(): expr and order match.\n",
            "INFO [2025-05-07 17:31:07] ::process_data:order_reduce:Reduction from positional data, new dimensions (r,c) = 12277,10309 Total=NA Min=NA Max=NA.\n",
            "INFO [2025-05-07 17:31:07] num genes removed taking into account provided gene ordering list: 477 = 3.88531400179197% removed.\n",
            "INFO [2025-05-07 17:31:07] -filtering out cells < 100 or > Inf, removing 73.5183 % of cells\n",
            "WARN [2025-05-07 17:31:08] Please use \"options(scipen = 100)\" before running infercnv if you are using the analysis_mode=\"subclusters\" option or you may encounter an error while the hclust is being generated.\n",
            "INFO [2025-05-07 17:31:08] validating infercnv_obj\n",
            "INFO [2025-05-07 17:31:08] ::process_data:Start\n",
            "INFO [2025-05-07 17:31:08] Creating output path infercnv_output\n",
            "INFO [2025-05-07 17:31:08] Checking for saved results.\n",
            "INFO [2025-05-07 17:31:08] \n",
            "\n",
            "\tSTEP 1: incoming data\n",
            "\n",
            "INFO [2025-05-07 17:31:13] \n",
            "\n",
            "\tSTEP 02: Removing lowly expressed genes\n",
            "\n",
            "INFO [2025-05-07 17:31:13] ::above_min_mean_expr_cutoff:Start\n",
            "INFO [2025-05-07 17:31:13] Removing 11323 genes from matrix as below mean expr threshold: 1\n",
            "INFO [2025-05-07 17:31:13] validating infercnv_obj\n",
            "INFO [2025-05-07 17:31:13] There are 477 genes and 2730 cells remaining in the expr matrix.\n",
            "INFO [2025-05-07 17:31:13] no genes removed due to min cells/gene filter\n",
            "INFO [2025-05-07 17:31:14] \n",
            "\n",
            "\tSTEP 03: normalization by sequencing depth\n",
            "\n",
            "INFO [2025-05-07 17:31:14] normalizing counts matrix by depth\n",
            "INFO [2025-05-07 17:31:14] Computed total sum normalization factor as median libsize: 1409.500000\n",
            "INFO [2025-05-07 17:31:14] Adding h-spike\n",
            "INFO [2025-05-07 17:31:14] -hspike modeling of B cell\n",
            "INFO [2025-05-07 17:31:45] -hspike modeling of CD4 T cell\n",
            "INFO [2025-05-07 17:32:16] validating infercnv_obj\n",
            "INFO [2025-05-07 17:32:16] normalizing counts matrix by depth\n",
            "INFO [2025-05-07 17:32:16] Using specified normalization factor: 1409.500000\n",
            "INFO [2025-05-07 17:32:17] \n",
            "\n",
            "\tSTEP 04: log transformation of data\n",
            "\n",
            "INFO [2025-05-07 17:32:17] transforming log2xplus1()\n",
            "INFO [2025-05-07 17:32:17] -mirroring for hspike\n",
            "INFO [2025-05-07 17:32:17] transforming log2xplus1()\n",
            "INFO [2025-05-07 17:32:20] \n",
            "\n",
            "\tSTEP 08: removing average of reference data (before smoothing)\n",
            "\n",
            "INFO [2025-05-07 17:32:20] ::subtract_ref_expr_from_obs:Start inv_log=FALSE, use_bounds=TRUE\n",
            "INFO [2025-05-07 17:32:20] subtracting mean(normal) per gene per cell across all data\n",
            "INFO [2025-05-07 17:32:20] -subtracting expr per gene, use_bounds=TRUE\n",
            "INFO [2025-05-07 17:32:20] -mirroring for hspike\n",
            "INFO [2025-05-07 17:32:20] ::subtract_ref_expr_from_obs:Start inv_log=FALSE, use_bounds=TRUE\n",
            "INFO [2025-05-07 17:32:20] subtracting mean(normal) per gene per cell across all data\n",
            "INFO [2025-05-07 17:32:21] -subtracting expr per gene, use_bounds=TRUE\n",
            "INFO [2025-05-07 17:32:24] \n",
            "\n",
            "\tSTEP 09: apply max centered expression threshold: 3\n",
            "\n",
            "INFO [2025-05-07 17:32:24] ::process_data:setting max centered expr, threshold set to: +/-:  3\n",
            "INFO [2025-05-07 17:32:24] -mirroring for hspike\n",
            "INFO [2025-05-07 17:32:24] ::process_data:setting max centered expr, threshold set to: +/-:  3\n",
            "INFO [2025-05-07 17:32:26] \n",
            "\n",
            "\tSTEP 10: Smoothing data per cell by chromosome\n",
            "\n",
            "INFO [2025-05-07 17:32:26] smooth_by_chromosome: chr: 1\n",
            "INFO [2025-05-07 17:32:28] smooth_by_chromosome: chr: 2\n",
            "INFO [2025-05-07 17:32:28] smooth_by_chromosome: chr: 3\n",
            "INFO [2025-05-07 17:32:29] smooth_by_chromosome: chr: 4\n",
            "INFO [2025-05-07 17:32:29] smooth_by_chromosome: chr: 5\n",
            "INFO [2025-05-07 17:32:30] smooth_by_chromosome: chr: 6\n",
            "INFO [2025-05-07 17:32:30] smooth_by_chromosome: chr: HSCHR6_MHC_MANN_CTG1\n",
            "INFO [2025-05-07 17:32:30] smooth_by_chromosome: chr: HSCHR6_MHC_QBL_CTG1\n",
            "INFO [2025-05-07 17:32:30] smooth_by_chromosome: chr: HSCHR6_1_CTG5\n",
            "INFO [2025-05-07 17:32:30] smooth_by_chromosome: chr: 7\n",
            "INFO [2025-05-07 17:32:31] smooth_by_chromosome: chr: HSCHR7_1_CTG6\n",
            "INFO [2025-05-07 17:32:31] smooth_by_chromosome: chr: X\n",
            "INFO [2025-05-07 17:32:31] smooth_by_chromosome: chr: 8\n",
            "INFO [2025-05-07 17:32:31] smooth_by_chromosome: chr: 9\n",
            "INFO [2025-05-07 17:32:32] smooth_by_chromosome: chr: 11\n",
            "INFO [2025-05-07 17:32:33] smooth_by_chromosome: chr: 10\n",
            "INFO [2025-05-07 17:32:33] smooth_by_chromosome: chr: 12\n",
            "INFO [2025-05-07 17:32:33] -mirroring for hspike\n",
            "INFO [2025-05-07 17:32:33] smooth_by_chromosome: chr: chrA\n",
            "INFO [2025-05-07 17:32:34] smooth_by_chromosome: chr: chr_0\n",
            "INFO [2025-05-07 17:32:34] smooth_by_chromosome: chr: chr_B\n",
            "INFO [2025-05-07 17:32:34] smooth_by_chromosome: chr: chr_0pt5\n",
            "INFO [2025-05-07 17:32:35] smooth_by_chromosome: chr: chr_C\n",
            "INFO [2025-05-07 17:32:35] smooth_by_chromosome: chr: chr_1pt5\n",
            "INFO [2025-05-07 17:32:36] smooth_by_chromosome: chr: chr_D\n",
            "INFO [2025-05-07 17:32:36] smooth_by_chromosome: chr: chr_2pt0\n",
            "INFO [2025-05-07 17:32:36] smooth_by_chromosome: chr: chr_E\n",
            "INFO [2025-05-07 17:32:37] smooth_by_chromosome: chr: chr_3pt0\n",
            "INFO [2025-05-07 17:32:37] smooth_by_chromosome: chr: chr_F\n",
            "INFO [2025-05-07 17:32:40] \n",
            "\n",
            "\tSTEP 11: re-centering data across chromosome after smoothing\n",
            "\n",
            "INFO [2025-05-07 17:32:40] ::center_smooth across chromosomes per cell\n",
            "INFO [2025-05-07 17:32:40] -mirroring for hspike\n",
            "INFO [2025-05-07 17:32:40] ::center_smooth across chromosomes per cell\n",
            "INFO [2025-05-07 17:32:43] \n",
            "\n",
            "\tSTEP 12: removing average of reference data (after smoothing)\n",
            "\n",
            "INFO [2025-05-07 17:32:43] ::subtract_ref_expr_from_obs:Start inv_log=FALSE, use_bounds=TRUE\n",
            "INFO [2025-05-07 17:32:43] subtracting mean(normal) per gene per cell across all data\n",
            "INFO [2025-05-07 17:32:43] -subtracting expr per gene, use_bounds=TRUE\n",
            "INFO [2025-05-07 17:32:43] -mirroring for hspike\n",
            "INFO [2025-05-07 17:32:43] ::subtract_ref_expr_from_obs:Start inv_log=FALSE, use_bounds=TRUE\n",
            "INFO [2025-05-07 17:32:43] subtracting mean(normal) per gene per cell across all data\n",
            "INFO [2025-05-07 17:32:44] -subtracting expr per gene, use_bounds=TRUE\n",
            "INFO [2025-05-07 17:32:47] \n",
            "\n",
            "\tSTEP 14: invert log2(FC) to FC\n",
            "\n",
            "INFO [2025-05-07 17:32:47] invert_log2(), computing 2^x\n",
            "INFO [2025-05-07 17:32:47] -mirroring for hspike\n",
            "INFO [2025-05-07 17:32:47] invert_log2(), computing 2^x\n",
            "INFO [2025-05-07 17:32:50] \n",
            "\n",
            "\tSTEP 15: computing tumor subclusters via leiden\n",
            "\n",
            "INFO [2025-05-07 17:32:50] define_signif_tumor_subclusters(p_val=0.1\n",
            "INFO [2025-05-07 17:32:50] define_signif_tumor_subclusters(), tumor: CD14 monocyte\n",
            "INFO [2025-05-07 17:32:50] Setting auto leiden resolution for CD14 monocyte to 0.0261178\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  COX8A, FERMT3, UBXN1, TRMT112, PRDX5, AHNAK, SF1, FTH1, FAU, TMEM258 \n",
            "\t   NEAT1, MS4A7, MALAT1, CFL1, MS4A6A, CTSW, CCDC85B, MPEG1, DRAP1, GRK2 \n",
            "\t   SELENOH, PPP1CA, SPI1, TBC1D10C, GSTP1, CD44, LAMTOR1, EIF3M, UCP2, LDHA \n",
            "Negative:  C1orf56, CDC42SE1, CTSS, S100A10, S100A11, MCL1, S100A9, TXNIP, S100A12, CSDE1 \n",
            "\t   S100A8, CAPZA1, S100A6, S100A4, RAP1A, JTB, TPM3, C1orf162, SSR2, ATP5PB \n",
            "\t   MNDA, IFI16, CD53, TAGLN2, CD48, LAMTOR5, ARHGAP30, FCER1G, PRPF38B, FCGR3A \n",
            "PC_ 2 \n",
            "Positive:  SF1, FAU, PRDX5, TRMT112, FERMT3, COX8A, NEAT1, MALAT1, UBXN1, CFL1 \n",
            "\t   AHNAK, CTSW, FTH1, CCDC85B, TMEM258, DRAP1, MS4A7, GRK2, MS4A6A, PPP1CA \n",
            "\t   MPEG1, TBC1D10C, SELENOH, GSTP1, SPI1, LAMTOR1, CD44, UCP2, EIF3M, SPCS2 \n",
            "Negative:  H2AFY, SKP1, TCF7, UQCRQ, IRF1, CDC42SE2, TGFBI, HINT1, HNRNPA0, SNX2 \n",
            "\t   PAIP2, TNFAIP8, UBE2D2, REEP5, CD14, CAST, NDFIP1, CSNK1A1, GLRX, CD74 \n",
            "\t   MEF2C, COX7C, ANXA6, VCAN, NPM1, TBCA, DUSP1, NSA2, ATP6V0E1, HIGD2A \n",
            "PC_ 3 \n",
            "Positive:  IRF1, CDC42SE2, UQCRQ, HINT1, TCF7, SNX2, SKP1, H2AFY, TGFBI, TNFAIP8 \n",
            "\t   HNRNPA0, REEP5, PAIP2, UBE2D2, CAST, CD14, GLRX, NDFIP1, MEF2C, CSNK1A1 \n",
            "\t   COX7C, CD74, VCAN, TBCA, ANXA6, NSA2, BTF3, NPM1, GZMA, DUSP1 \n",
            "Negative:  COX5B, ACTR3, IGKC, GNLY, DBI, VAMP8, MZT2B, TGOLN2, TMSB10, CXCR4 \n",
            "\t   ZEB2, MOB1A, PRPF40A, PCBP1, NR4A2, PLEK, GCA, PPIG, ACTR2, WIPF1 \n",
            "\t   RTN4, ATP5MC3, HNRNPA3, CALM2, STK17B, ZFP36L2, ANKRD44, SF3B1, COX7A2L, CFLAR \n",
            "PC_ 4 \n",
            "Positive:  CXCR4, ZEB2, MZT2B, PRPF40A, DBI, NR4A2, ACTR3, COX5B, GCA, IGKC \n",
            "\t   PPIG, GNLY, WIPF1, VAMP8, ATP5MC3, TGOLN2, HNRNPA3, STK17B, ANKRD44, TMSB10 \n",
            "\t   SF3B1, CFLAR, MOB1A, EEF1B2, XRCC5, PCBP1, ARPC2, PLEK, SP110, ACTR2 \n",
            "Negative:  MYO1G, IKZF1, SEC61G, PPIA, CHCHD2, DBNL, NCF1, SEPT7, FGL2, CPVL \n",
            "\t   CD36, SEM1, CBX3, BRI3, HNRNPA2B1, ARPC1B, CYCS, TOMM7, ATP5MF, NDUFA4 \n",
            "\t   RAC1, ACTB, LAMTOR4, GNB2, KMT2E, NAMPT, CAPZA2, ATP6V1F, MTPN, NDUFB2 \n",
            "PC_ 5 \n",
            "Positive:  DAZAP2, BIN2, EIF4B, TMBIM6, PFDN5, TUBA1A, PRR13, TUBA1B, PCBP2, ATP5MC2 \n",
            "\t   SLC38A2, HNRNPA1, SCAF11, NCKAP1L, LDHB, BLOC1S1, CD63, ARHGDIB, MYL6, PLBD1 \n",
            "\t   ATP5F1B, YBX3, PTGES3, CLEC7A, NACA, CD69, USP15, CLEC2D, TMBIM4, RAP1B \n",
            "Negative:  PRELID1, HNRNPH1, HIGD2A, ATP6V0E1, CANX, DUSP1, RNF130, NPM1, RACK1, ANXA6 \n",
            "\t   CD74, CSNK1A1, NDFIP1, CD14, UBE2D2, PAIP2, HNRNPA0, TGFBI, H2AFY, SKP1 \n",
            "\t   TCF7, UQCRQ, IRF1, CDC42SE2, HINT1, SNX2, TNFAIP8, REEP5, CAST, GLRX \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-05-07 17:32:51] define_signif_tumor_subclusters(), tumor: CD8 T cell\n",
            "INFO [2025-05-07 17:32:51] Setting auto leiden resolution for CD8 T cell to 0.0757741\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  CSDE1, CAPZA1, RAP1A, TXNIP, C1orf162, ATP5PB, MCL1, CD53, CTSS, LAMTOR5 \n",
            "\t   C1orf56, CDC42SE1, PRPF38B, S100A10, SELENOF, S100A11, GNG5, S100A9, SRSF11, S100A12 \n",
            "\t   SERBP1, S100A8, HNRNPA0, PAIP2, CD14, TGFBI, UBE2D2, NDFIP1, JAK1, H2AFY \n",
            "Negative:  PPP1CA, GRK2, TBC1D10C, DRAP1, CCDC85B, GSTP1, CTSW, CFL1, LAMTOR1, UCP2 \n",
            "\t   MALAT1, SPCS2, NEAT1, NDUFC2, CTSC, FAU, TMEM123, SF1, CASP4, PRDX5 \n",
            "\t   CASP1, TRMT112, CARD16, ATM, FERMT3, PCSK7, COX8A, JAML, UBXN1, CD3E \n",
            "PC_ 2 \n",
            "Positive:  RAP1A, CAPZA1, C1orf162, ATP5PB, CSDE1, CD53, TXNIP, LAMTOR5, MCL1, PRPF38B \n",
            "\t   CTSS, SELENOF, C1orf56, GNG5, CDC42SE1, SRSF11, S100A10, SERBP1, JAK1, S100A11 \n",
            "\t   JUN, S100A9, TMEM59, UQCRH, S100A12, PRDX1, S100A8, ATP6V0B, YBX1, S100A6 \n",
            "Negative:  SNX2, TNFAIP8, HINT1, REEP5, CDC42SE2, CAST, GLRX, IRF1, MEF2C, TBCA \n",
            "\t   NSA2, VCAN, COX7C, BTF3, UQCRQ, TCF7, SKP1, H2AFY, GZMA, TGFBI \n",
            "\t   HNRNPA0, FYB1, PAIP2, UBE2D2, IL7R, CD14, SUB1, NDFIP1, NDUFS6, CSNK1A1 \n",
            "PC_ 3 \n",
            "Positive:  FTH1, TMEM258, AHNAK, MS4A7, MS4A6A, UBXN1, MPEG1, SELENOH, COX8A, SPI1 \n",
            "\t   FERMT3, CD44, TRMT112, EIF3M, PRDX5, LDHA, SF1, ATP5MC2, HNRNPA1, PCBP2 \n",
            "\t   NCKAP1L, PRR13, PFDN5, FAU, C11orf58, EIF4B, BLOC1S1, BIN2, CD63, DAZAP2 \n",
            "Negative:  VAMP8, TGOLN2, GNLY, TMSB10, IGKC, MOB1A, PCBP1, COX5B, PLEK, ACTR2 \n",
            "\t   ACTR3, RTN4, DBI, CALM2, MZT2B, CXCR4, ZFP36L2, ZEB2, COX7A2L, PRPF40A \n",
            "\t   NR4A2, SRSF7, GCA, PPIG, WIPF1, CDC42EP3, ATP5MC3, YPEL5, HNRNPA3, STK17B \n",
            "PC_ 4 \n",
            "Positive:  MTPN, NDUFB2, ATP6V1F, TRBC1, TRBC2, GSTK1, CAPZA2, GIMAP1, GIMAP7, GIMAP4 \n",
            "\t   NAMPT, KMT2E, GNB2, LAMTOR4, FGL2, CD36, NCF1, ATP5MF, SEM1, ARPC1B \n",
            "\t   BRI3, CHCHD2, SEC61G, IKZF1, MYO1G, PPIA, DBNL, TCF7, UQCRQ, SKP1 \n",
            "Negative:  CLEC7A, YBX3, CD69, CLEC2D, PLBD1, KLRB1, M6PR, ARHGDIB, SLC2A3, PHB2 \n",
            "\t   LDHB, PTPN6, SCAF11, C12orf57, SLC38A2, TPI1, GAPDH, TUBA1B, TUBA1A, TMBIM6 \n",
            "\t   DAZAP2, BIN2, EIF4B, PFDN5, PRR13, PCBP2, ATP5MC2, HNRNPA1, NCKAP1L, BLOC1S1 \n",
            "PC_ 5 \n",
            "Positive:  MYO1G, PPIA, IKZF1, SEC61G, DBNL, CHCHD2, NCF1, FGL2, CD36, SEM1 \n",
            "\t   SEPT7, BRI3, ARPC1B, CPVL, ATP5MF, CBX3, LAMTOR4, HNRNPA2B1, CYCS, GNB2 \n",
            "\t   KMT2E, TOMM7, NDUFA4, NAMPT, RAC1, ACTB, CAPZA2, ATP6V1F, MTPN, NDUFB2 \n",
            "Negative:  UQCRQ, IRF1, TCF7, CDC42SE2, HINT1, SKP1, SNX2, H2AFY, TNFAIP8, REEP5 \n",
            "\t   TGFBI, CAST, GLRX, HNRNPA0, MEF2C, PAIP2, UBE2D2, COX7C, CD14, VCAN \n",
            "\t   NDFIP1, TBCA, BTF3, GZMA, NSA2, FYB1, CSNK1A1, IL7R, CD74, NPM1 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-05-07 17:32:51] define_signif_tumor_subclusters(), tumor: Dendritic\n",
            "INFO [2025-05-07 17:32:51] Setting auto leiden resolution for Dendritic to 0.363072\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  MZT2B, CXCR4, ACTR3, SF3B6, COX5B, DBI, RHOB, HADHA, IGKC, ID2 \n",
            "\t   GNLY, ZEB2, OST4, YPEL5, CDC42EP3, SRSF7, PRPF40A, VAMP8, NR4A2, COX7A2L \n",
            "\t   GCA, PPIG, TGOLN2, ZFP36L2, TMSB10, WIPF1, MOB1A, CALM2, ATP5MC3, PCBP1 \n",
            "Negative:  S100A11, S100A9, S100A10, CDC42SE1, S100A12, C1orf56, S100A8, CTSS, S100A6, S100A4 \n",
            "\t   MCL1, JTB, TXNIP, TPM3, CSDE1, SSR2, MNDA, CAPZA1, IFI16, RAP1A \n",
            "\t   TAGLN2, C1orf162, CD48, ARHGAP30, ATP5PB, FCER1G, FCGR3A, CD53, RCSD1, LAMTOR5 \n",
            "PC_ 2 \n",
            "Positive:  RNH1, TALDO1, IFITM2, POLR2L, CTSD, LSP1, RHOG, EIF3F, EIF4G2, PSMA1 \n",
            "\t   LDHA, EIF3M, C11orf58, CD44, SPI1, SELENOH, MPEG1, MS4A6A, MS4A7, TMEM258 \n",
            "\t   FTH1, AHNAK, UBXN1, COX8A, FERMT3, TRMT112, PRDX5, SF1, FAU, NEAT1 \n",
            "Negative:  SNX2, TNFAIP8, HINT1, REEP5, CDC42SE2, CAST, GLRX, IRF1, MEF2C, COX7C \n",
            "\t   UQCRQ, VCAN, TBCA, TCF7, NSA2, SKP1, H2AFY, BTF3, TGFBI, GZMA \n",
            "\t   FYB1, IL7R, HNRNPA0, SUB1, PAIP2, NDUFS6, UBE2D2, CD14, NDFIP1, CSNK1A1 \n",
            "PC_ 3 \n",
            "Positive:  ENO1, PARK7, TNFRSF1B, EFHD2, GNB1, CAPZB, MINOS1, CDC42, SRRM1, SYF2 \n",
            "\t   RSRP1, TMEM50A, SH3BGRL3, CD52, HMGN2, WASF2, FGR, ATP5IF1, LAPTM5, PTP4A2 \n",
            "\t   LCK, SFPQ, CSF3R, NDUFS5, CAP1, PPT1, SMAP2, YBX1, ATP6V0B, PRDX1 \n",
            "Negative:  APLP2, HSPA8, ATP5MG, CD3G, CD3D, CD3E, JAML, PCSK7, ATM, CARD16 \n",
            "\t   CASP1, CASP4, TMEM123, CTSC, NDUFC2, SPCS2, UCP2, LAMTOR1, GSTP1, TBC1D10C \n",
            "\t   PPP1CA, GRK2, DRAP1, CCDC85B, CTSW, CFL1, MALAT1, NEAT1, FAU, SF1 \n",
            "PC_ 4 \n",
            "Positive:  EIF4B, BIN2, DAZAP2, PFDN5, TMBIM6, PRR13, TUBA1A, PCBP2, TUBA1B, ATP5MC2 \n",
            "\t   SLC38A2, SCAF11, HNRNPA1, LDHB, ARHGDIB, PLBD1, NCKAP1L, YBX3, CLEC7A, BLOC1S1 \n",
            "\t   CD69, CLEC2D, CD63, KLRB1, M6PR, MYL6, SLC2A3, ATP5F1B, PHB2, GAPDH \n",
            "Negative:  COPS9, LRRFIP1, PLP2, ARL4C, RBM3, UXT, PTMA, CFP, TIMP1, DDX3X \n",
            "\t   NDUFB11, NCL, ATP6AP2, SEC61B, KLF4, ZFAND5, HNRNPK, ANP32B, TLE4, CYBB \n",
            "\t   OSTF1, TXN, CLTA, ANXA1, ATP6V1G1, TLN1, HSPA5, UBE2R2, EIF2S3, SP100 \n",
            "PC_ 5 \n",
            "Positive:  RNASET2, BCLAF1, ARID1B, MARCKS, SF3B5, ABRACL, FYN, CD164, SNX3, PNISR \n",
            "\t   PNRC1, COX7A2, EEF1A1, CLIC1, RIPOR2, HSP90AB1, DEK, PSMB8, TAPBP, CUTA \n",
            "\t   PSMB9, LY86, CNPY3, BRD2, SRSF3, SERPINB1, CCND3, CLEC7A, YBX3, CD69 \n",
            "Negative:  CMTM6, BRK1, NKTR, TMA7, ARPC4, RHOA, GNAI2, SPCS1, TKT, ARL6IP5 \n",
            "\t   FOXP1, NFKBIZ, CD47, NDUFB4, HCLS1, HNRNPU, LYST, ARID4B, GIMAP1, TOMM20 \n",
            "\t   GIMAP4, GIMAP7, GSTK1, GUK1, TRBC2, ARF1, TRBC1, H3F3A, SRP9, NDUFB2 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-05-07 17:32:52] define_signif_tumor_subclusters(), tumor: FCGR3A monocyte\n",
            "INFO [2025-05-07 17:32:52] Setting auto leiden resolution for FCGR3A monocyte to 0.202492\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  SF1, PRDX5, TRMT112, CCDC85B, FAU, CTSW, CFL1, DRAP1, FERMT3, MALAT1 \n",
            "\t   GRK2, NEAT1, COX8A, PPP1CA, UBXN1, TBC1D10C, GSTP1, AHNAK, FTH1, LAMTOR1 \n",
            "\t   TMEM258, MS4A7, MS4A6A, UCP2, MPEG1, SELENOH, SPCS2, SPI1, CD44, EIF3M \n",
            "Negative:  SELENOF, PRPF38B, GNG5, LAMTOR5, SRSF11, SERBP1, JAK1, CD53, JUN, TMEM59 \n",
            "\t   UQCRH, ATP5PB, PRDX1, C1orf162, RAP1A, ATP6V0B, CAPZA1, CSDE1, YBX1, TXNIP \n",
            "\t   SMAP2, MCL1, CTSS, PPT1, C1orf56, CDC42SE1, CAP1, S100A10, S100A11, NDUFS5 \n",
            "PC_ 2 \n",
            "Positive:  GNLY, IGKC, VAMP8, TGOLN2, COX5B, TMSB10, ACTR3, MOB1A, PCBP1, DBI \n",
            "\t   PLEK, ACTR2, CALM2, RTN4, MZT2B, ZFP36L2, COX7A2L, CXCR4, SRSF7, ZEB2 \n",
            "\t   CDC42EP3, PRPF40A, NR4A2, YPEL5, GCA, PPIG, OST4, WIPF1, HADHA, ATP5MC3 \n",
            "Negative:  DAZAP2, TMBIM6, BIN2, TUBA1A, SLC38A2, TUBA1B, SCAF11, EIF4B, LDHB, PFDN5 \n",
            "\t   ARHGDIB, PRR13, PLBD1, PCBP2, YBX3, ATP5MC2, HNRNPA1, CLEC7A, NCKAP1L, BLOC1S1 \n",
            "\t   CD63, CD69, MYL6, ATP5F1B, CLEC2D, PTGES3, NACA, KLRB1, USP15, TMBIM4 \n",
            "PC_ 3 \n",
            "Positive:  FCER1G, FCGR3A, ARHGAP30, RCSD1, CD48, SELL, PRRC2C, TAGLN2, IER5, NCF2 \n",
            "\t   IFI16, ARPC5, MNDA, RGS2, SSR2, TPM3, PTPRC, JTB, BTG2, S100A4 \n",
            "\t   MDM4, S100A6, NUCKS1, TRAF3IP3, S100A8, SRP9, S100A12, H3F3A, S100A9, S100A11 \n",
            "Negative:  SEC61G, CHCHD2, IKZF1, NCF1, FGL2, MYO1G, CD36, PPIA, DBNL, SEPT7 \n",
            "\t   CPVL, SEM1, CBX3, BRI3, HNRNPA2B1, TOMM7, CYCS, NDUFA4, ARPC1B, RAC1 \n",
            "\t   ACTB, ATP5MF, LAMTOR4, GNB2, KMT2E, NAMPT, CAPZA2, ATP6V1F, MTPN, NDUFB2 \n",
            "PC_ 4 \n",
            "Positive:  EDF1, FCN1, FNBP1, SET, C9orf16, HSPA5, TIMP1, NDUFB11, ATP6V1G1, DDX3X \n",
            "\t   CYBB, EIF2S3, ATP6AP2, TXN, CFP, SAT1, KLF4, SEC61B, UXT, SH3KBP1 \n",
            "\t   ANP32B, HNRNPK, RBM3, AP1S2, PLP2, TLE4, TMSB4X, OSTF1, CD99, SLC25A6 \n",
            "Negative:  GIMAP1, GIMAP4, GIMAP7, GSTK1, TRBC2, ATP5MF, MTPN, TRBC1, ARPC1B, KMT2E \n",
            "\t   NAMPT, BRI3, LAMTOR4, GNB2, NDUFB2, CAPZA2, ATP6V1F, SEM1, CD36, FGL2 \n",
            "\t   NCF1, CHCHD2, SEC61G, IKZF1, MYO1G, PPIA, DBNL, SEPT7, CPVL, CBX3 \n",
            "PC_ 5 \n",
            "Positive:  RIPOR2, CLIC1, DEK, LY86, PSMB8, SERPINB1, PSMB9, BRD2, TAPBP, CUTA \n",
            "\t   SRSF3, CCND3, CNPY3, RNASET2, COX7A2, HSP90AB1, EEF1A1, PNRC1, ARID1B, ABRACL \n",
            "\t   SF3B5, PNISR, BCLAF1, SNX3, CD164, MARCKS, FYN, CTSB, ASAH1, GNB1 \n",
            "Negative:  PCBP1, PLEK, MOB1A, ACTR2, TMSB10, TGOLN2, RTN4, VAMP8, CALM2, ZFP36L2 \n",
            "\t   GNLY, COX7A2L, SRSF7, IGKC, CDC42EP3, YPEL5, OST4, HADHA, COX5B, SF3B6 \n",
            "\t   RHOB, ID2, ACTR3, DBI, MZT2B, CXCR4, ZEB2, PRPF40A, NR4A2, GCA \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-05-07 17:32:52] define_signif_tumor_subclusters(), tumor: Megakaryocyte\n",
            "INFO [2025-05-07 17:32:52] Less cells in group Megakaryocyte than k_nn setting. Keeping as a single subcluster.\n",
            "INFO [2025-05-07 17:32:52] define_signif_tumor_subclusters(), tumor: NK cell\n",
            "INFO [2025-05-07 17:32:52] Setting auto leiden resolution for NK cell to 0.107509\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  BLOC1S1, NCKAP1L, HNRNPA1, CD63, ATP5MC2, PCBP2, MYL6, PRR13, ATP5F1B, PFDN5 \n",
            "\t   PTGES3, EIF4B, NACA, BIN2, USP15, DAZAP2, TMBIM4, TMBIM6, RAP1B, TUBA1A \n",
            "\t   TUBA1B, SLC38A2, SCAF11, LDHB, ARHGDIB, PLBD1, YBX3, CLEC7A, CD69, CLEC2D \n",
            "Negative:  CTSS, C1orf56, MCL1, CDC42SE1, S100A10, TXNIP, CSDE1, S100A11, CAPZA1, S100A9 \n",
            "\t   RAP1A, S100A12, C1orf162, S100A8, ATP5PB, S100A6, CD53, S100A4, LAMTOR5, JTB \n",
            "\t   PRPF38B, TPM3, SELENOF, SSR2, GNG5, MNDA, SRSF11, IFI16, SERBP1, TAGLN2 \n",
            "PC_ 2 \n",
            "Positive:  NPM1, DUSP1, ANXA6, ATP6V0E1, CD74, HIGD2A, CSNK1A1, HNRNPH1, RNF130, CANX \n",
            "\t   RACK1, PRELID1, NDFIP1, CD14, UBE2D2, PNRC1, PAIP2, PNISR, HNRNPA0, COX7A2 \n",
            "\t   TGFBI, SNX3, EEF1A1, H2AFY, FYN, CD164, MARCKS, SKP1, BCLAF1, ABRACL \n",
            "Negative:  FAU, SF1, NEAT1, PRDX5, MALAT1, TRMT112, CFL1, FERMT3, COX8A, CTSW \n",
            "\t   UBXN1, CCDC85B, AHNAK, DRAP1, FTH1, GRK2, TMEM258, PPP1CA, MS4A7, TBC1D10C \n",
            "\t   MS4A6A, GSTP1, MPEG1, LAMTOR1, SELENOH, UCP2, SPI1, SPCS2, CD44, NDUFC2 \n",
            "PC_ 3 \n",
            "Positive:  CAP1, PPT1, NDUFS5, SMAP2, CSF3R, YBX1, ATP6V0B, SFPQ, PRDX1, LCK \n",
            "\t   UQCRH, PTP4A2, TMEM59, LAPTM5, JUN, ATP5IF1, FGR, JAK1, WASF2, SERBP1 \n",
            "\t   HMGN2, SRSF11, CD52, GNG5, SH3BGRL3, TMEM50A, SELENOF, RSRP1, PRPF38B, SYF2 \n",
            "Negative:  PPIG, WIPF1, ATP5MC3, HNRNPA3, GCA, STK17B, NR4A2, ANKRD44, SF3B1, PRPF40A \n",
            "\t   CFLAR, ZEB2, EEF1B2, CXCR4, MZT2B, XRCC5, DBI, ACTR3, ARPC2, COX5B \n",
            "\t   IGKC, SP110, GNLY, SP100, VAMP8, TGOLN2, NCL, TMSB10, PTMA, MOB1A \n",
            "PC_ 4 \n",
            "Positive:  GUK1, ARF1, TOMM20, ARID4B, LYST, HNRNPU, H3F3A, SRP9, TRAF3IP3, NUCKS1 \n",
            "\t   MDM4, BTG2, PTPRC, RGS2, ARPC5, RNASET2, ABRACL, ARID1B, BCLAF1, SF3B5 \n",
            "\t   DEK, NCF2, RIPOR2, CLIC1, LY86, MARCKS, SERPINB1, PSMB8, FYN, CUTA \n",
            "Negative:  TGFBI, HNRNPA0, H2AFY, PAIP2, SKP1, UBE2D2, TCF7, CD14, UQCRQ, NDFIP1 \n",
            "\t   IRF1, CSNK1A1, CDC42SE2, HINT1, CD74, ANXA6, SNX2, NPM1, TNFAIP8, DUSP1 \n",
            "\t   REEP5, ATP6V0E1, CAST, HIGD2A, GLRX, PRELID1, MEF2C, HNRNPH1, COX7C, CANX \n",
            "PC_ 5 \n",
            "Positive:  NUCKS1, PTPRC, TRAF3IP3, RGS2, MDM4, BTG2, ARPC5, SRP9, NCF2, IER5 \n",
            "\t   H3F3A, SPCS1, GNAI2, PRRC2C, RHOA, TKT, TMA7, NKTR, CMTM6, ARF1 \n",
            "\t   BRK1, ARL6IP5, SELL, ARPC4, GUK1, FOXP1, TOMM20, KLF6, RCSD1, NFKBIZ \n",
            "Negative:  PLBD1, ARHGDIB, YBX3, LDHB, CLEC7A, SCAF11, PHB2, SLC2A3, M6PR, CD69 \n",
            "\t   SLC38A2, PTPN6, KLRB1, CLEC2D, C12orf57, TPI1, GAPDH, TUBA1B, TUBA1A, TMBIM6 \n",
            "\t   DAZAP2, BIN2, EIF4B, PFDN5, PRR13, PCBP2, ATP5MC2, HNRNPA1, NCKAP1L, BLOC1S1 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-05-07 17:32:53] define_signif_tumor_subclusters(), tumor: B cell\n",
            "INFO [2025-05-07 17:32:53] Setting auto leiden resolution for B cell to 0.0504175\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  C1orf162, RAP1A, ATP5PB, CAPZA1, CD53, LAMTOR5, CSDE1, PRPF38B, TXNIP, SELENOF \n",
            "\t   MCL1, GNG5, CTSS, SRSF11, C1orf56, SERBP1, CDC42SE1, JAK1, S100A10, JUN \n",
            "\t   S100A11, TMEM59, S100A9, UQCRH, S100A12, PRDX1, ATP6V0B, S100A8, YBX1, S100A6 \n",
            "Negative:  PRDX5, FERMT3, TRMT112, FAU, COX8A, UBXN1, SF1, NEAT1, MALAT1, AHNAK \n",
            "\t   CFL1, CTSW, FTH1, CCDC85B, TMEM258, DRAP1, MS4A7, GRK2, PPP1CA, TBC1D10C \n",
            "\t   GSTP1, MS4A6A, LAMTOR1, UCP2, MPEG1, SPCS2, NDUFC2, SELENOH, CTSC, SPI1 \n",
            "PC_ 2 \n",
            "Positive:  MALAT1, CFL1, NEAT1, CTSW, CCDC85B, FAU, DRAP1, SF1, GRK2, PRDX5 \n",
            "\t   PPP1CA, TRMT112, TBC1D10C, FERMT3, COX8A, GSTP1, UBXN1, LAMTOR1, AHNAK, UCP2 \n",
            "\t   FTH1, SPCS2, TMEM258, NDUFC2, MS4A7, CTSC, MS4A6A, MPEG1, TMEM123, SELENOH \n",
            "Negative:  TGFBI, H2AFY, HNRNPA0, SKP1, TCF7, UQCRQ, PAIP2, IRF1, CDC42SE2, UBE2D2 \n",
            "\t   HINT1, CD14, SNX2, NDFIP1, TNFAIP8, CSNK1A1, CD74, REEP5, ANXA6, CAST \n",
            "\t   GLRX, NPM1, MEF2C, DUSP1, COX7C, ATP6V0E1, VCAN, HIGD2A, TBCA, PRELID1 \n",
            "PC_ 3 \n",
            "Positive:  VAMP8, TGOLN2, GNLY, IGKC, TMSB10, COX5B, ACTR3, MOB1A, DBI, PCBP1 \n",
            "\t   MZT2B, PLEK, CXCR4, ZEB2, ACTR2, PRPF40A, RTN4, NR4A2, GCA, CALM2 \n",
            "\t   PPIG, ZFP36L2, WIPF1, COX7A2L, ATP5MC3, SRSF7, HNRNPA3, CDC42EP3, STK17B, YPEL5 \n",
            "Negative:  TRMT112, FERMT3, PRDX5, COX8A, SF1, UBXN1, FAU, AHNAK, FTH1, NEAT1 \n",
            "\t   TMEM258, MALAT1, CFL1, CTSW, CCDC85B, MS4A7, DRAP1, GRK2, MS4A6A, PPP1CA \n",
            "\t   MPEG1, TBC1D10C, SELENOH, GSTP1, SPI1, LAMTOR1, CD44, EIF3M, LDHA, UCP2 \n",
            "PC_ 4 \n",
            "Positive:  LAMTOR4, ATP5MF, GNB2, ARPC1B, KMT2E, NAMPT, CAPZA2, BRI3, ATP6V1F, MTPN \n",
            "\t   SEM1, NDUFB2, CD36, TRBC1, TRBC2, FGL2, GSTK1, GIMAP7, NCF1, GIMAP4 \n",
            "\t   GIMAP1, CHCHD2, SEC61G, IKZF1, MYO1G, PPIA, DBNL, SEPT7, CPVL, CBX3 \n",
            "Negative:  SLC38A2, TUBA1A, TUBA1B, TMBIM6, SCAF11, LDHB, DAZAP2, ARHGDIB, BIN2, PLBD1 \n",
            "\t   EIF4B, YBX3, PFDN5, CLEC7A, PRR13, CD69, PCBP2, ATP5MC2, CLEC2D, HNRNPA1 \n",
            "\t   KLRB1, NCKAP1L, M6PR, BLOC1S1, SLC2A3, CD63, MYL6, PHB2, ATP5F1B, PTPN6 \n",
            "PC_ 5 \n",
            "Positive:  SKP1, TCF7, H2AFY, TGFBI, UQCRQ, UBE2D2, HNRNPA0, PAIP2, IRF1, CD14 \n",
            "\t   NDFIP1, CDC42SE2, CSNK1A1, HINT1, CD74, SNX2, ANXA6, TNFAIP8, NPM1, REEP5 \n",
            "\t   CAST, DUSP1, GLRX, ATP6V0E1, MEF2C, HIGD2A, COX7C, PRELID1, VCAN, HNRNPH1 \n",
            "Negative:  SCAF11, LDHB, ARHGDIB, SLC38A2, TUBA1B, PLBD1, TUBA1A, YBX3, TMBIM6, DAZAP2 \n",
            "\t   CLEC7A, TPI1, CLEC2D, BIN2, GAPDH, C12orf57, CD69, KLRB1, PTPN6, M6PR \n",
            "\t   PHB2, SLC2A3, EIF4B, PFDN5, PRR13, PCBP2, ATP5MC2, HNRNPA1, NCKAP1L, BLOC1S1 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-05-07 17:32:53] define_signif_tumor_subclusters(), tumor: CD4 T cell\n",
            "INFO [2025-05-07 17:32:53] Setting auto leiden resolution for CD4 T cell to 0.0229312\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Warning message:\n",
            "“Data is of class matrix. Coercing to dgCMatrix.”\n",
            "Finding variable features for layer counts\n",
            "\n",
            "Centering and scaling data matrix\n",
            "\n",
            "PC_ 1 \n",
            "Positive:  RAP1A, C1orf162, CAPZA1, ATP5PB, CSDE1, CD53, TXNIP, LAMTOR5, MCL1, PRPF38B \n",
            "\t   CTSS, SELENOF, C1orf56, GNG5, CDC42SE1, SRSF11, S100A10, SERBP1, S100A11, JAK1 \n",
            "\t   S100A9, JUN, S100A12, TMEM59, S100A8, UQCRH, S100A6, PRDX1, S100A4, ATP6V0B \n",
            "Negative:  PAIP2, UBE2D2, CD14, NDFIP1, HNRNPA0, CSNK1A1, TGFBI, H2AFY, CD74, SKP1 \n",
            "\t   ANXA6, NPM1, TCF7, UQCRQ, DUSP1, IRF1, ATP6V0E1, CDC42SE2, HINT1, HIGD2A \n",
            "\t   SNX2, TNFAIP8, PRELID1, REEP5, HNRNPH1, RACK1, RNF130, CANX, CAST, GLRX \n",
            "PC_ 2 \n",
            "Positive:  TMSB10, MOB1A, TGOLN2, PCBP1, PLEK, ACTR2, VAMP8, RTN4, CALM2, GNLY \n",
            "\t   ZFP36L2, IGKC, COX5B, COX7A2L, ACTR3, DBI, SRSF7, MZT2B, CDC42EP3, CXCR4 \n",
            "\t   ZEB2, YPEL5, PRPF40A, OST4, NR4A2, HADHA, GCA, PPIG, SF3B6, WIPF1 \n",
            "Negative:  SF1, PRDX5, FAU, TRMT112, NEAT1, FERMT3, MALAT1, COX8A, CFL1, CTSW \n",
            "\t   UBXN1, CCDC85B, AHNAK, DRAP1, FTH1, GRK2, TMEM258, PPP1CA, MS4A7, TBC1D10C \n",
            "\t   MS4A6A, GSTP1, LAMTOR1, MPEG1, UCP2, SELENOH, SPCS2, SPI1, NDUFC2, CD44 \n",
            "PC_ 3 \n",
            "Positive:  ATP5MC2, HNRNPA1, PCBP2, PRR13, NCKAP1L, PFDN5, BLOC1S1, EIF4B, CD63, BIN2 \n",
            "\t   DAZAP2, MYL6, TMBIM6, ATP5F1B, TUBA1A, PTGES3, TUBA1B, NACA, SLC38A2, USP15 \n",
            "\t   SCAF11, TMBIM4, LDHB, RAP1B, ARHGDIB, PLBD1, YBX3, CLEC7A, CD69, CLEC2D \n",
            "Negative:  ACTR3, COX5B, DBI, IGKC, GNLY, MZT2B, CXCR4, VAMP8, ZEB2, PRPF40A \n",
            "\t   TGOLN2, NR4A2, TMSB10, GCA, MOB1A, PCBP1, PPIG, PLEK, WIPF1, ACTR2 \n",
            "\t   ATP5MC3, HNRNPA3, RTN4, STK17B, CALM2, ANKRD44, ZFP36L2, SF3B1, COX7A2L, CFLAR \n",
            "PC_ 4 \n",
            "Positive:  NCF1, FGL2, CD36, CHCHD2, SEM1, BRI3, ARPC1B, SEC61G, ATP5MF, IKZF1 \n",
            "\t   LAMTOR4, MYO1G, GNB2, PPIA, KMT2E, DBNL, NAMPT, SEPT7, CAPZA2, CPVL \n",
            "\t   ATP6V1F, CBX3, MTPN, HNRNPA2B1, NDUFB2, CYCS, TOMM7, TRBC1, NDUFA4, RAC1 \n",
            "Negative:  IRF1, UQCRQ, TCF7, CDC42SE2, SKP1, TGFBI, H2AFY, HINT1, HNRNPA0, SNX2 \n",
            "\t   PAIP2, TNFAIP8, UBE2D2, CD14, REEP5, NDFIP1, CAST, CSNK1A1, GLRX, CD74 \n",
            "\t   MEF2C, ANXA6, NPM1, COX7C, DUSP1, VCAN, ATP6V0E1, TBCA, HIGD2A, PRELID1 \n",
            "PC_ 5 \n",
            "Positive:  SLC38A2, SCAF11, TUBA1B, TUBA1A, LDHB, TMBIM6, ARHGDIB, DAZAP2, BIN2, PLBD1 \n",
            "\t   YBX3, EIF4B, CLEC7A, PFDN5, CD69, PRR13, CLEC2D, PCBP2, KLRB1, ATP5MC2 \n",
            "\t   HNRNPA1, M6PR, NCKAP1L, SLC2A3, BLOC1S1, PHB2, CD63, PTPN6, MYL6, ATP5F1B \n",
            "Negative:  CD36, SEM1, FGL2, BRI3, NCF1, ARPC1B, CHCHD2, ATP5MF, SEC61G, LAMTOR4 \n",
            "\t   IKZF1, GNB2, MYO1G, KMT2E, PPIA, NAMPT, DBNL, CAPZA2, SEPT7, ATP6V1F \n",
            "\t   CPVL, CBX3, MTPN, HNRNPA2B1, NDUFB2, CYCS, TRBC1, TOMM7, TRBC2, NDUFA4 \n",
            "\n",
            "Computing nearest neighbor graph\n",
            "\n",
            "Computing SNN\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO [2025-05-07 17:32:54] -mirroring for hspike\n",
            "INFO [2025-05-07 17:32:54] define_signif_tumor_subclusters(p_val=0.1\n",
            "INFO [2025-05-07 17:32:54] define_signif_tumor_subclusters(), tumor: spike_tumor_cell_B cell\n",
            "INFO [2025-05-07 17:32:54] cut tree into: 1 groups\n",
            "INFO [2025-05-07 17:32:54] -processing spike_tumor_cell_B cell,spike_tumor_cell_B cell_s1\n",
            "INFO [2025-05-07 17:32:54] define_signif_tumor_subclusters(), tumor: spike_tumor_cell_CD4 T cell\n",
            "INFO [2025-05-07 17:32:54] cut tree into: 1 groups\n",
            "INFO [2025-05-07 17:32:54] -processing spike_tumor_cell_CD4 T cell,spike_tumor_cell_CD4 T cell_s1\n",
            "INFO [2025-05-07 17:32:54] define_signif_tumor_subclusters(), tumor: simnorm_cell_B cell\n",
            "INFO [2025-05-07 17:32:54] cut tree into: 1 groups\n",
            "INFO [2025-05-07 17:32:54] -processing simnorm_cell_B cell,simnorm_cell_B cell_s1\n",
            "INFO [2025-05-07 17:32:54] define_signif_tumor_subclusters(), tumor: simnorm_cell_CD4 T cell\n",
            "INFO [2025-05-07 17:32:54] cut tree into: 1 groups\n",
            "INFO [2025-05-07 17:32:54] -processing simnorm_cell_CD4 T cell,simnorm_cell_CD4 T cell_s1\n",
            "INFO [2025-05-07 17:32:57] ::plot_cnv:Start\n",
            "INFO [2025-05-07 17:32:57] ::plot_cnv:Current data dimensions (r,c)=477,2730 Total=1311338.48246305 Min=0.311386364997398 Max=9.12526404188546.\n",
            "INFO [2025-05-07 17:32:57] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
            "INFO [2025-05-07 17:32:57] plot_cnv(): auto thresholding at: (0.728505 , 1.285515)\n",
            "INFO [2025-05-07 17:32:57] plot_cnv_observation:Start\n",
            "INFO [2025-05-07 17:32:57] Observation data size: Cells= 1367 Genes= 477\n",
            "INFO [2025-05-07 17:32:57] clustering observations via method: ward.D\n",
            "INFO [2025-05-07 17:32:57] Number of cells in group(1) is 100\n",
            "INFO [2025-05-07 17:32:57] group size being clustered:  100,477\n",
            "INFO [2025-05-07 17:32:57] Number of cells in group(2) is 81\n",
            "INFO [2025-05-07 17:32:57] group size being clustered:  81,477\n",
            "INFO [2025-05-07 17:32:57] Number of cells in group(3) is 65\n",
            "INFO [2025-05-07 17:32:57] group size being clustered:  65,477\n",
            "INFO [2025-05-07 17:32:57] Number of cells in group(4) is 64\n",
            "INFO [2025-05-07 17:32:57] group size being clustered:  64,477\n",
            "INFO [2025-05-07 17:32:57] Number of cells in group(5) is 63\n",
            "INFO [2025-05-07 17:32:57] group size being clustered:  63,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(6) is 59\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  59,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(7) is 56\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  56,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(8) is 56\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  56,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(9) is 54\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  54,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(10) is 51\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  51,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(11) is 41\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  41,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(12) is 35\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  35,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(13) is 25\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  25,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(14) is 25\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  25,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(15) is 19\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  19,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(16) is 13\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  13,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(17) is 12\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  12,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(18) is 10\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  10,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(19) is 4\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  4,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(20) is 2\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  2,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(21) is 1\n",
            "INFO [2025-05-07 17:32:58] Skipping group: 21, since less than 2 entries\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(22) is 1\n",
            "INFO [2025-05-07 17:32:58] Skipping group: 22, since less than 2 entries\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(23) is 105\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  105,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(24) is 46\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  46,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(25) is 42\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  42,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(26) is 23\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  23,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(27) is 13\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  13,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(28) is 7\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  7,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(29) is 4\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  4,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(30) is 1\n",
            "INFO [2025-05-07 17:32:58] Skipping group: 30, since less than 2 entries\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(31) is 1\n",
            "INFO [2025-05-07 17:32:58] Skipping group: 31, since less than 2 entries\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(32) is 39\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  39,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(33) is 40\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  40,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(34) is 36\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  36,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(35) is 1\n",
            "INFO [2025-05-07 17:32:58] Skipping group: 35, since less than 2 entries\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(36) is 11\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  11,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(37) is 96\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  96,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(38) is 20\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  20,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(39) is 17\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  17,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(40) is 14\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  14,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(41) is 9\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  9,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(42) is 2\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  2,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(43) is 2\n",
            "INFO [2025-05-07 17:32:58] group size being clustered:  2,477\n",
            "INFO [2025-05-07 17:32:58] Number of cells in group(44) is 1\n",
            "INFO [2025-05-07 17:32:58] Skipping group: 44, since less than 2 entries\n",
            "INFO [2025-05-07 17:32:58] plot_cnv_observation:Writing observation groupings/color.\n",
            "INFO [2025-05-07 17:32:58] plot_cnv_observation:Done writing observation groupings/color.\n",
            "INFO [2025-05-07 17:32:58] plot_cnv_observation:Writing observation heatmap thresholds.\n",
            "INFO [2025-05-07 17:32:58] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
            "INFO [2025-05-07 17:32:58] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-05-07 17:32:58] Quantiles of plotted data range: 0.728505212190201,0.944247807945654,1,1.06039666242174,1.28551477300889\n",
            "INFO [2025-05-07 17:32:59] plot_cnv_references:Start\n",
            "INFO [2025-05-07 17:32:59] Reference data size: Cells= 1363 Genes= 477\n",
            "INFO [2025-05-07 17:32:59] plot_cnv_references:Number reference groups= 36\n",
            "INFO [2025-05-07 17:32:59] plot_cnv_references:Plotting heatmap.\n",
            "INFO [2025-05-07 17:32:59] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-05-07 17:32:59] Quantiles of plotted data range: 0.728505212190201,0.966160181622641,1,1.03516004406222,1.28551477300889\n",
            "INFO [2025-05-07 17:33:02] ::plot_cnv:Start\n",
            "INFO [2025-05-07 17:33:02] ::plot_cnv:Current data dimensions (r,c)=477,2730 Total=1311338.48246305 Min=0.311386364997398 Max=9.12526404188546.\n",
            "INFO [2025-05-07 17:33:02] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
            "INFO [2025-05-07 17:33:02] plot_cnv(): auto thresholding at: (0.728505 , 1.285515)\n",
            "INFO [2025-05-07 17:33:02] plot_cnv_observation:Start\n",
            "INFO [2025-05-07 17:33:02] Observation data size: Cells= 1367 Genes= 477\n",
            "INFO [2025-05-07 17:33:02] plot_cnv_observation:Writing observation groupings/color.\n",
            "INFO [2025-05-07 17:33:02] plot_cnv_observation:Done writing observation groupings/color.\n",
            "INFO [2025-05-07 17:33:02] plot_cnv_observation:Writing observation heatmap thresholds.\n",
            "INFO [2025-05-07 17:33:02] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
            "INFO [2025-05-07 17:33:02] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-05-07 17:33:02] Quantiles of plotted data range: 0.728505212190201,0.944247807945654,1,1.06039666242174,1.28551477300889\n",
            "INFO [2025-05-07 17:33:03] plot_cnv_references:Start\n",
            "INFO [2025-05-07 17:33:03] Reference data size: Cells= 1363 Genes= 477\n",
            "INFO [2025-05-07 17:33:03] plot_cnv_references:Number reference groups= 2\n",
            "INFO [2025-05-07 17:33:03] plot_cnv_references:Plotting heatmap.\n",
            "INFO [2025-05-07 17:33:03] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-05-07 17:33:03] Quantiles of plotted data range: 0.728505212190201,0.966160181622641,1,1.03516004406222,1.28551477300889\n",
            "INFO [2025-05-07 17:33:04] \n",
            "\n",
            "\tSTEP 17: HMM-based CNV prediction\n",
            "\n",
            "INFO [2025-05-07 17:33:04] predict_CNV_via_HMM_on_tumor_subclusters\n",
            "INFO [2025-05-07 17:33:09] -done predicting CNV based on initial tumor subclusters\n",
            "INFO [2025-05-07 17:33:09] get_predicted_CNV_regions(subcluster)\n",
            "INFO [2025-05-07 17:33:09] -processing cell_group_name: CD14 monocyte.CD14 monocyte_s13, size: 100\n",
            "INFO [2025-05-07 17:33:10] -processing cell_group_name: CD14 monocyte.CD14 monocyte_s11, size: 81\n",
            "INFO [2025-05-07 17:33:10] -processing cell_group_name: CD14 monocyte.CD14 monocyte_s10, size: 65\n",
            "INFO [2025-05-07 17:33:11] -processing cell_group_name: CD14 monocyte.CD14 monocyte_s1, size: 64\n",
            "INFO [2025-05-07 17:33:11] -processing cell_group_name: CD14 monocyte.CD14 monocyte_s8, size: 63\n",
            "INFO [2025-05-07 17:33:11] -processing cell_group_name: CD14 monocyte.CD14 monocyte_s7, size: 59\n",
            "INFO [2025-05-07 17:33:12] -processing cell_group_name: CD14 monocyte.CD14 monocyte_s3, size: 56\n",
            "INFO [2025-05-07 17:33:12] -processing cell_group_name: CD14 monocyte.CD14 monocyte_s14, size: 56\n",
            "INFO [2025-05-07 17:33:13] -processing cell_group_name: CD14 monocyte.CD14 monocyte_s4, size: 54\n",
            "INFO [2025-05-07 17:33:13] -processing cell_group_name: CD14 monocyte.CD14 monocyte_s9, size: 51\n",
            "INFO [2025-05-07 17:33:13] -processing cell_group_name: CD14 monocyte.CD14 monocyte_s15, size: 41\n",
            "INFO [2025-05-07 17:33:14] -processing cell_group_name: CD14 monocyte.CD14 monocyte_s2, size: 35\n",
            "INFO [2025-05-07 17:33:14] -processing cell_group_name: CD14 monocyte.CD14 monocyte_s12, size: 25\n",
            "INFO [2025-05-07 17:33:14] -processing cell_group_name: CD14 monocyte.CD14 monocyte_s17, size: 25\n",
            "INFO [2025-05-07 17:33:15] -processing cell_group_name: CD14 monocyte.CD14 monocyte_s5, size: 19\n",
            "INFO [2025-05-07 17:33:15] -processing cell_group_name: CD14 monocyte.CD14 monocyte_s16, size: 13\n",
            "INFO [2025-05-07 17:33:15] -processing cell_group_name: CD14 monocyte.CD14 monocyte_s6, size: 12\n",
            "INFO [2025-05-07 17:33:16] -processing cell_group_name: CD14 monocyte.CD14 monocyte_s18, size: 10\n",
            "INFO [2025-05-07 17:33:16] -processing cell_group_name: CD14 monocyte.CD14 monocyte_s20, size: 4\n",
            "INFO [2025-05-07 17:33:16] -processing cell_group_name: CD14 monocyte.CD14 monocyte_s21, size: 2\n",
            "INFO [2025-05-07 17:33:17] -processing cell_group_name: CD14 monocyte.CD14 monocyte_s19, size: 1\n",
            "INFO [2025-05-07 17:33:17] -processing cell_group_name: CD14 monocyte.CD14 monocyte_s22, size: 1\n",
            "INFO [2025-05-07 17:33:17] -processing cell_group_name: CD8 T cell.CD8 T cell_s2, size: 105\n",
            "INFO [2025-05-07 17:33:18] -processing cell_group_name: CD8 T cell.CD8 T cell_s3, size: 46\n",
            "INFO [2025-05-07 17:33:18] -processing cell_group_name: CD8 T cell.CD8 T cell_s5, size: 42\n",
            "INFO [2025-05-07 17:33:18] -processing cell_group_name: CD8 T cell.CD8 T cell_s1, size: 23\n",
            "INFO [2025-05-07 17:33:19] -processing cell_group_name: CD8 T cell.CD8 T cell_s4, size: 13\n",
            "INFO [2025-05-07 17:33:19] -processing cell_group_name: CD8 T cell.CD8 T cell_s7, size: 7\n",
            "INFO [2025-05-07 17:33:19] -processing cell_group_name: CD8 T cell.CD8 T cell_s6, size: 4\n",
            "INFO [2025-05-07 17:33:20] -processing cell_group_name: CD8 T cell.CD8 T cell_s8, size: 1\n",
            "INFO [2025-05-07 17:33:20] -processing cell_group_name: CD8 T cell.CD8 T cell_s9, size: 1\n",
            "INFO [2025-05-07 17:33:20] -processing cell_group_name: Dendritic.Dendritic_s1, size: 39\n",
            "INFO [2025-05-07 17:33:21] -processing cell_group_name: FCGR3A monocyte.FCGR3A monocyte_s2, size: 40\n",
            "INFO [2025-05-07 17:33:21] -processing cell_group_name: FCGR3A monocyte.FCGR3A monocyte_s1, size: 36\n",
            "INFO [2025-05-07 17:33:22] -processing cell_group_name: FCGR3A monocyte.FCGR3A monocyte_s3, size: 1\n",
            "INFO [2025-05-07 17:33:22] -processing cell_group_name: Megakaryocyte.Megakaryocyte, size: 11\n",
            "INFO [2025-05-07 17:33:22] -processing cell_group_name: NK cell.NK cell_s1, size: 96\n",
            "INFO [2025-05-07 17:33:23] -processing cell_group_name: NK cell.NK cell_s2, size: 20\n",
            "INFO [2025-05-07 17:33:23] -processing cell_group_name: NK cell.NK cell_s4, size: 17\n",
            "INFO [2025-05-07 17:33:23] -processing cell_group_name: NK cell.NK cell_s3, size: 14\n",
            "INFO [2025-05-07 17:33:24] -processing cell_group_name: NK cell.NK cell_s5, size: 9\n",
            "INFO [2025-05-07 17:33:24] -processing cell_group_name: NK cell.NK cell_s6, size: 2\n",
            "INFO [2025-05-07 17:33:24] -processing cell_group_name: NK cell.NK cell_s7, size: 2\n",
            "INFO [2025-05-07 17:33:25] -processing cell_group_name: NK cell.NK cell_s8, size: 1\n",
            "INFO [2025-05-07 17:33:25] -processing cell_group_name: B cell.B cell_s2, size: 74\n",
            "INFO [2025-05-07 17:33:25] -processing cell_group_name: B cell.B cell_s1, size: 68\n",
            "INFO [2025-05-07 17:33:26] -processing cell_group_name: B cell.B cell_s9, size: 49\n",
            "INFO [2025-05-07 17:33:26] -processing cell_group_name: B cell.B cell_s5, size: 44\n",
            "INFO [2025-05-07 17:33:26] -processing cell_group_name: B cell.B cell_s3, size: 42\n",
            "INFO [2025-05-07 17:33:27] -processing cell_group_name: B cell.B cell_s4, size: 40\n",
            "INFO [2025-05-07 17:33:27] -processing cell_group_name: B cell.B cell_s6, size: 29\n",
            "INFO [2025-05-07 17:33:28] -processing cell_group_name: B cell.B cell_s8, size: 21\n",
            "INFO [2025-05-07 17:33:28] -processing cell_group_name: B cell.B cell_s11, size: 10\n",
            "INFO [2025-05-07 17:33:28] -processing cell_group_name: B cell.B cell_s7, size: 5\n",
            "INFO [2025-05-07 17:33:29] -processing cell_group_name: B cell.B cell_s12, size: 3\n",
            "INFO [2025-05-07 17:33:29] -processing cell_group_name: B cell.B cell_s14, size: 2\n",
            "INFO [2025-05-07 17:33:29] -processing cell_group_name: B cell.B cell_s10, size: 1\n",
            "INFO [2025-05-07 17:33:30] -processing cell_group_name: B cell.B cell_s13, size: 1\n",
            "INFO [2025-05-07 17:33:30] -processing cell_group_name: CD4 T cell.CD4 T cell_s3, size: 127\n",
            "INFO [2025-05-07 17:33:30] -processing cell_group_name: CD4 T cell.CD4 T cell_s1, size: 110\n",
            "INFO [2025-05-07 17:33:31] -processing cell_group_name: CD4 T cell.CD4 T cell_s12, size: 82\n",
            "INFO [2025-05-07 17:33:31] -processing cell_group_name: CD4 T cell.CD4 T cell_s2, size: 81\n",
            "INFO [2025-05-07 17:33:31] -processing cell_group_name: CD4 T cell.CD4 T cell_s10, size: 73\n",
            "INFO [2025-05-07 17:33:32] -processing cell_group_name: CD4 T cell.CD4 T cell_s16, size: 70\n",
            "INFO [2025-05-07 17:33:32] -processing cell_group_name: CD4 T cell.CD4 T cell_s11, size: 66\n",
            "INFO [2025-05-07 17:33:33] -processing cell_group_name: CD4 T cell.CD4 T cell_s6, size: 53\n",
            "INFO [2025-05-07 17:33:33] -processing cell_group_name: CD4 T cell.CD4 T cell_s8, size: 51\n",
            "INFO [2025-05-07 17:33:33] -processing cell_group_name: CD4 T cell.CD4 T cell_s15, size: 49\n",
            "INFO [2025-05-07 17:33:34] -processing cell_group_name: CD4 T cell.CD4 T cell_s9, size: 44\n",
            "INFO [2025-05-07 17:33:34] -processing cell_group_name: CD4 T cell.CD4 T cell_s13, size: 42\n",
            "INFO [2025-05-07 17:33:34] -processing cell_group_name: CD4 T cell.CD4 T cell_s14, size: 38\n",
            "INFO [2025-05-07 17:33:35] -processing cell_group_name: CD4 T cell.CD4 T cell_s17, size: 34\n",
            "INFO [2025-05-07 17:33:35] -processing cell_group_name: CD4 T cell.CD4 T cell_s7, size: 29\n",
            "INFO [2025-05-07 17:33:35] -processing cell_group_name: CD4 T cell.CD4 T cell_s4, size: 10\n",
            "INFO [2025-05-07 17:33:37] -processing cell_group_name: CD4 T cell.CD4 T cell_s19, size: 4\n",
            "INFO [2025-05-07 17:33:37] -processing cell_group_name: CD4 T cell.CD4 T cell_s21, size: 4\n",
            "INFO [2025-05-07 17:33:37] -processing cell_group_name: CD4 T cell.CD4 T cell_s18, size: 3\n",
            "INFO [2025-05-07 17:33:38] -processing cell_group_name: CD4 T cell.CD4 T cell_s20, size: 2\n",
            "INFO [2025-05-07 17:33:38] -processing cell_group_name: CD4 T cell.CD4 T cell_s5, size: 1\n",
            "INFO [2025-05-07 17:33:38] -processing cell_group_name: CD4 T cell.CD4 T cell_s22, size: 1\n",
            "INFO [2025-05-07 17:33:39] -writing cell clusters file: infercnv_output/17_HMM_predHMMi6.leiden.hmm_mode-subclusters.cell_groupings\n",
            "INFO [2025-05-07 17:33:39] -writing cnv regions file: infercnv_output/17_HMM_predHMMi6.leiden.hmm_mode-subclusters.pred_cnv_regions.dat\n",
            "INFO [2025-05-07 17:33:39] -writing per-gene cnv report: infercnv_output/17_HMM_predHMMi6.leiden.hmm_mode-subclusters.pred_cnv_genes.dat\n",
            "INFO [2025-05-07 17:33:39] -writing gene ordering info: infercnv_output/17_HMM_predHMMi6.leiden.hmm_mode-subclusters.genes_used.dat\n",
            "INFO [2025-05-07 17:33:42] ::plot_cnv:Start\n",
            "INFO [2025-05-07 17:33:42] ::plot_cnv:Current data dimensions (r,c)=477,2730 Total=4062416 Min=1 Max=6.\n",
            "INFO [2025-05-07 17:33:42] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
            "INFO [2025-05-07 17:33:42] plot_cnv_observation:Start\n",
            "INFO [2025-05-07 17:33:42] Observation data size: Cells= 1367 Genes= 477\n",
            "INFO [2025-05-07 17:33:42] plot_cnv_observation:Writing observation groupings/color.\n",
            "INFO [2025-05-07 17:33:42] plot_cnv_observation:Done writing observation groupings/color.\n",
            "INFO [2025-05-07 17:33:42] plot_cnv_observation:Writing observation heatmap thresholds.\n",
            "INFO [2025-05-07 17:33:42] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
            "INFO [2025-05-07 17:33:42] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-05-07 17:33:42] Quantiles of plotted data range: 1,2,3,4,6\n",
            "INFO [2025-05-07 17:33:42] plot_cnv_references:Start\n",
            "INFO [2025-05-07 17:33:42] Reference data size: Cells= 1363 Genes= 477\n",
            "INFO [2025-05-07 17:33:43] plot_cnv_references:Number reference groups= 2\n",
            "INFO [2025-05-07 17:33:43] plot_cnv_references:Plotting heatmap.\n",
            "INFO [2025-05-07 17:33:43] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-05-07 17:33:43] Quantiles of plotted data range: 1,3,3,3,6\n",
            "INFO [2025-05-07 17:33:43] \n",
            "\n",
            "\tSTEP 18: Run Bayesian Network Model on HMM predicted CNVs\n",
            "\n",
            "INFO [2025-05-07 17:33:43] Creating the following Directory:  infercnv_output/BayesNetOutput.HMMi6.leiden.hmm_mode-subclusters\n",
            "INFO [2025-05-07 17:33:43] Initializing new MCM InferCNV Object.\n",
            "INFO [2025-05-07 17:33:43] validating infercnv_obj\n",
            "INFO [2025-05-07 17:33:43] Total CNV's:  811\n",
            "INFO [2025-05-07 17:33:43] Loading BUGS Model.\n",
            "INFO [2025-05-07 17:33:44] Running Sampling Using Parallel with  4 Cores\n",
            "INFO [2025-05-07 17:39:38] Obtaining probabilities post-sampling\n",
            "INFO [2025-05-07 17:41:05] Gibbs sampling time:  7.35831303993861  Minutes\n",
            "INFO [2025-05-07 17:41:19] ::plot_cnv:Start\n",
            "INFO [2025-05-07 17:41:19] ::plot_cnv:Current data dimensions (r,c)=477,2730 Total=511921.890542421 Min=0 Max=0.99138363885186.\n",
            "INFO [2025-05-07 17:41:19] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
            "INFO [2025-05-07 17:41:19] plot_cnv_observation:Start\n",
            "INFO [2025-05-07 17:41:19] Observation data size: Cells= 1367 Genes= 477\n",
            "INFO [2025-05-07 17:41:19] plot_cnv_observation:Writing observation groupings/color.\n",
            "INFO [2025-05-07 17:41:19] plot_cnv_observation:Done writing observation groupings/color.\n",
            "INFO [2025-05-07 17:41:19] plot_cnv_observation:Writing observation heatmap thresholds.\n",
            "INFO [2025-05-07 17:41:19] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
            "INFO [2025-05-07 17:41:19] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-05-07 17:41:19] Quantiles of plotted data range: 0,0,0.769692534369413,0.922996952074611,0.9906421637069\n",
            "INFO [2025-05-07 17:41:19] plot_cnv_references:Start\n",
            "INFO [2025-05-07 17:41:19] Reference data size: Cells= 1363 Genes= 477\n",
            "INFO [2025-05-07 17:41:20] plot_cnv_references:Number reference groups= 2\n",
            "INFO [2025-05-07 17:41:20] plot_cnv_references:Plotting heatmap.\n",
            "INFO [2025-05-07 17:41:20] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-05-07 17:41:20] Quantiles of plotted data range: 0,0,0,0.659287037831219,0.99138363885186\n",
            "INFO [2025-05-07 17:41:32] \n",
            "\n",
            "\tSTEP 19: Filter HMM predicted CNVs based on the Bayesian Network Model results and BayesMaxPNormal\n",
            "\n",
            "INFO [2025-05-07 17:41:32] Attempting to removing CNV(s) with a probability of being normal above  0.5\n",
            "INFO [2025-05-07 17:41:32] Removing  1  CNV(s) identified by the HMM.\n",
            "INFO [2025-05-07 17:41:32] Total CNV's after removing:  810\n",
            "INFO [2025-05-07 17:41:32] Reassigning CNVs based on state probabilities.\n",
            "INFO [2025-05-07 17:41:32] Changing the following CNV's states assigned by the HMM to the following based on the CNV's state probabilities.\n",
            " 8-region_17 : 2  (P= 0.248212431874748 ) ->  3 (P= 0.326346574558711 )\n",
            "12-region_23 : 2  (P= 0.358636775618884 ) ->  1 (P= 0.510418482940435 )\n",
            "8-region_38 : 2  (P= 0.226249406475979 ) ->  3 (P= 0.375787819436637 )\n",
            "9-region_39 : 4  (P= 0.28098103126819 ) ->  5 (P= 0.317928528934078 )\n",
            "3-region_51 : 2  (P= 0.22514578006079 ) ->  3 (P= 0.381070794000654 )\n",
            "X-region_58 : 5  (P= 0.253519126284847 ) ->  4 (P= 0.293009217190908 )\n",
            "X-region_59 : 4  (P= 0.225972028957271 ) ->  5 (P= 0.267957624840686 )\n",
            "10-region_66 : 5  (P= 0.326202981230813 ) ->  6 (P= 0.441825191952522 )\n",
            "12-region_67 : 2  (P= 0.243694419690802 ) ->  1 (P= 0.544281121277108 )\n",
            "2-region_73 : 2  (P= 0.270791246647288 ) ->  1 (P= 0.472630701238576 )\n",
            "10-region_86 : 5  (P= 0.267886892218888 ) ->  6 (P= 0.390692180624416 )\n",
            "12-region_87 : 2  (P= 0.285906882405933 ) ->  1 (P= 0.527674181600889 )\n",
            "8-region_102 : 2  (P= 0.243983838301941 ) ->  1 (P= 0.275795394760126 )\n",
            "2-region_113 : 2  (P= 0.37010088328739 ) ->  1 (P= 0.430732135023792 )\n",
            "X-region_121 : 5  (P= 0.273063692021131 ) ->  4 (P= 0.315933462803165 )\n",
            "8-region_124 : 2  (P= 0.262267693459908 ) ->  3 (P= 0.353300644312076 )\n",
            "12-region_129 : 2  (P= 0.261343252504213 ) ->  3 (P= 0.355242404838563 )\n",
            "7-region_139 : 4  (P= 0.251851453755233 ) ->  3 (P= 0.484707428763522 )\n",
            "8-region_143 : 2  (P= 0.209465060839004 ) ->  1 (P= 0.387630727995657 )\n",
            "12-region_148 : 2  (P= 0.286616586339637 ) ->  1 (P= 0.519670454416563 )\n",
            "8-region_162 : 2  (P= 0.274991988484692 ) ->  1 (P= 0.29054094300679 )\n",
            "9-region_163 : 5  (P= 0.25830483608479 ) ->  6 (P= 0.322692333518463 )\n",
            "11-region_164 : 4  (P= 0.241328971779425 ) ->  5 (P= 0.388302705414106 )\n",
            "12-region_167 : 2  (P= 0.195679623149352 ) ->  1 (P= 0.53193404641159 )\n",
            "2-region_172 : 2  (P= 0.290092354525719 ) ->  1 (P= 0.361548114797255 )\n",
            "5-region_176 : 2  (P= 0.266707760612016 ) ->  1 (P= 0.483169605968422 )\n",
            "X-region_180 : 4  (P= 0.182226159128663 ) ->  5 (P= 0.320188780593428 )\n",
            "8-region_183 : 2  (P= 0.316171481675107 ) ->  1 (P= 0.33279073312863 )\n",
            "10-region_189 : 5  (P= 0.29981609962423 ) ->  6 (P= 0.417249955637187 )\n",
            "1-region_195 : 5  (P= 0.363552742561411 ) ->  6 (P= 0.407750755700531 )\n",
            "X-region_204 : 4  (P= 0.223130305161377 ) ->  3 (P= 0.30405646610494 )\n",
            "1-region_212 : 2  (P= 0.342727297703449 ) ->  3 (P= 0.402415860004272 )\n",
            "X-region_224 : 4  (P= 0.214794476608625 ) ->  5 (P= 0.294784295755447 )\n",
            "8-region_227 : 2  (P= 0.105526958414967 ) ->  3 (P= 0.403083436416449 )\n",
            "9-region_228 : 5  (P= 0.233090770878706 ) ->  6 (P= 0.341200290606502 )\n",
            "8-region_249 : 2  (P= 0.243162100356288 ) ->  3 (P= 0.39163912971571 )\n",
            "11-region_253 : 2  (P= 0.323405178388811 ) ->  3 (P= 0.41450840231013 )\n",
            "5-region_263 : 2  (P= 0.290647815017181 ) ->  1 (P= 0.578793534049522 )\n",
            "8-region_269 : 2  (P= 0.243388274428191 ) ->  1 (P= 0.304309055693032 )\n",
            "9-region_270 : 5  (P= 0.354371992347852 ) ->  6 (P= 0.354744037717706 )\n",
            "10-region_274 : 5  (P= 0.383529382695112 ) ->  6 (P= 0.424381185695338 )\n",
            "12-region_275 : 2  (P= 0.35599790152581 ) ->  1 (P= 0.418581296777472 )\n",
            "2-region_282 : 2  (P= 0.256981500832725 ) ->  1 (P= 0.389560611437009 )\n",
            "X-region_290 : 4  (P= 0.194049554660087 ) ->  5 (P= 0.35393638153259 )\n",
            "9-region_293 : 4  (P= 0.255131149107968 ) ->  5 (P= 0.260207541419287 )\n",
            "8-region_312 : 2  (P= 0.239693579898638 ) ->  1 (P= 0.24134021682028 )\n",
            "9-region_313 : 5  (P= 0.280878991275484 ) ->  4 (P= 0.281238324561478 )\n",
            "10-region_316 : 5  (P= 0.239419796061507 ) ->  6 (P= 0.400848617096158 )\n",
            "11-region_334 : 4  (P= 0.263181922717229 ) ->  3 (P= 0.3170593569162 )\n",
            "10-region_336 : 5  (P= 0.210690414205727 ) ->  6 (P= 0.369817556199897 )\n",
            "8-region_352 : 4  (P= 0.165700244019147 ) ->  3 (P= 0.224314335151422 )\n",
            "9-region_354 : 5  (P= 0.219424770723815 ) ->  6 (P= 0.390384099251715 )\n",
            "1-region_359 : 2  (P= 0.271857141795041 ) ->  3 (P= 0.352546992241126 )\n",
            "X-region_372 : 4  (P= 0.187169475528865 ) ->  3 (P= 0.375083237061676 )\n",
            "9-region_376 : 5  (P= 0.313204502368351 ) ->  6 (P= 0.313774947019458 )\n",
            "11-region_378 : 4  (P= 0.248698419956185 ) ->  5 (P= 0.310849007418279 )\n",
            "3-region_387 : 4  (P= 0.188596253076599 ) ->  5 (P= 0.200719949997904 )\n",
            "X-region_394 : 4  (P= 0.100637188421156 ) ->  6 (P= 0.298072591518374 )\n",
            "8-region_396 : 2  (P= 0.200711468225216 ) ->  3 (P= 0.299876987448022 )\n",
            "1-region_403 : 4  (P= 0.125186440639039 ) ->  5 (P= 0.251302645591919 )\n",
            "1-region_406 : 5  (P= 0.126896576116821 ) ->  6 (P= 0.250902064081873 )\n",
            "2-region_407 : 2  (P= 0.125247772369897 ) ->  1 (P= 0.250344352258785 )\n",
            "5-region_410 : 2  (P= 0.248550765085805 ) ->  1 (P= 0.251518298450773 )\n",
            "6-region_411 : 2  (P= 0.125247772369897 ) ->  1 (P= 0.250344352258785 )\n",
            "X-region_414 : 5  (P= 0.249766298074945 ) ->  6 (P= 0.252454386571452 )\n",
            "11-region_417 : 4  (P= 0.125617645325798 ) ->  5 (P= 0.249627236054989 )\n",
            "10-region_420 : 5  (P= 0.248934120892649 ) ->  6 (P= 0.249259973067212 )\n",
            "3-region_469 : 2  (P= 0.206706776654936 ) ->  1 (P= 0.379861969235575 )\n",
            "4-region_470 : 2  (P= 0.159225862843451 ) ->  1 (P= 0.404345634897285 )\n",
            "8-region_476 : 2  (P= 0.144192114330715 ) ->  1 (P= 0.350486865209321 )\n",
            "11-region_479 : 4  (P= 0.282488251081768 ) ->  3 (P= 0.486067857455832 )\n",
            "10-region_480 : 4  (P= 0.180100071365153 ) ->  3 (P= 0.379057197779906 )\n",
            "1-region_483 : 4  (P= 0.222472093982787 ) ->  5 (P= 0.345226878930095 )\n",
            "2-region_485 : 2  (P= 0.270388901545404 ) ->  3 (P= 0.458023581612316 )\n",
            "3-region_486 : 2  (P= 0.230924104112271 ) ->  1 (P= 0.308464476348252 )\n",
            "8-region_493 : 2  (P= 0.169479938183625 ) ->  1 (P= 0.383690107181152 )\n",
            "11-region_496 : 4  (P= 0.327627988307541 ) ->  3 (P= 0.345505629237641 )\n",
            "12-region_499 : 2  (P= 0.0972926084940755 ) ->  1 (P= 0.364672660422178 )\n",
            "3-region_503 : 2  (P= 0.124232120758684 ) ->  3 (P= 0.333320181066004 )\n",
            "4-region_504 : 2  (P= 0.145143882153434 ) ->  1 (P= 0.416240409542554 )\n",
            "5-region_506 : 2  (P= 0.269791773738114 ) ->  1 (P= 0.312687759671736 )\n",
            "X-region_510 : 2  (P= 0.249839796886345 ) ->  1 (P= 0.292143591494071 )\n",
            "8-region_511 : 2  (P= 0.186479907883182 ) ->  1 (P= 0.375533767017914 )\n",
            "9-region_512 : 5  (P= 0.169039468136109 ) ->  6 (P= 0.270586807793973 )\n",
            "3-region_519 : 2  (P= 0.103217440809943 ) ->  1 (P= 0.450626581416132 )\n",
            "4-region_520 : 2  (P= 0.146461489734771 ) ->  1 (P= 0.368931321678107 )\n",
            "6-region_522 : 2  (P= 0.138902384747097 ) ->  3 (P= 0.414035220190379 )\n",
            "7-region_524 : 2  (P= 0.207121268126206 ) ->  3 (P= 0.415168690686189 )\n",
            "9-region_527 : 4  (P= 0.137988113285823 ) ->  3 (P= 0.241977981411609 )\n",
            "1-region_532 : 4  (P= 0.3152780481015 ) ->  3 (P= 0.315495612543885 )\n",
            "6-region_538 : 2  (P= 0.0530567558443985 ) ->  1 (P= 0.370054245354804 )\n",
            "10-region_547 : 4  (P= 0.105490136916692 ) ->  5 (P= 0.2623016420337 )\n",
            "12-region_549 : 2  (P= 0.158965668517673 ) ->  1 (P= 0.314482440319753 )\n",
            "1-region_550 : 4  (P= 0.22807833611609 ) ->  5 (P= 0.309606808135655 )\n",
            "2-region_552 : 4  (P= 0.230281398039101 ) ->  3 (P= 0.307575945506547 )\n",
            "3-region_554 : 1  (P= 0.305524526894642 ) ->  2 (P= 0.385856416874122 )\n",
            "4-region_555 : 2  (P= 0.154147181563903 ) ->  1 (P= 0.232790809031637 )\n",
            "7-region_560 : 4  (P= 0.0765359184120221 ) ->  3 (P= 0.30717846055595 )\n",
            "X-region_561 : 2  (P= 0.0789335230174792 ) ->  1 (P= 0.310850178313637 )\n",
            "9-region_563 : 2  (P= 0.0758736981873847 ) ->  3 (P= 0.462016891472298 )\n",
            "10-region_566 : 4  (P= 0.15448288917309 ) ->  3 (P= 0.38576637346894 )\n",
            "2-region_570 : 4  (P= 0.099492637587773 ) ->  3 (P= 0.300413127253634 )\n",
            "3-region_571 : 1  (P= 0.296041450245092 ) ->  2 (P= 0.301267777945816 )\n",
            "11-region_581 : 4  (P= 0.100272212317516 ) ->  5 (P= 0.301184320739657 )\n",
            "10-region_584 : 4  (P= 0.198668025377175 ) ->  3 (P= 0.298633023387156 )\n",
            "3-region_627 : 2  (P= 0.17670981396684 ) ->  1 (P= 0.401703605006162 )\n",
            "5-region_629 : 2  (P= 0.311859262159006 ) ->  1 (P= 0.355376516886427 )\n",
            "X-region_635 : 5  (P= 0.289005522750146 ) ->  4 (P= 0.311325863718811 )\n",
            "9-region_637 : 4  (P= 0.243829580642784 ) ->  3 (P= 0.334801705711141 )\n",
            "10-region_640 : 5  (P= 0.298472934711873 ) ->  6 (P= 0.332828020159593 )\n",
            "2-region_644 : 2  (P= 0.263557788050966 ) ->  1 (P= 0.387397400610382 )\n",
            "8-region_654 : 2  (P= 0.217981626488365 ) ->  3 (P= 0.309661819358347 )\n",
            "9-region_656 : 2  (P= 0.154662074764264 ) ->  1 (P= 0.323912526717341 )\n",
            "10-region_660 : 4  (P= 0.260577289703662 ) ->  3 (P= 0.347119060085669 )\n",
            "12-region_661 : 2  (P= 0.239875911727907 ) ->  1 (P= 0.542582915684125 )\n",
            "2-region_664 : 2  (P= 0.261393534619293 ) ->  1 (P= 0.40618648771196 )\n",
            "X-region_672 : 4  (P= 0.286947377792635 ) ->  3 (P= 0.287068448919373 )\n",
            "8-region_674 : 2  (P= 0.167235917198011 ) ->  3 (P= 0.356152314841111 )\n",
            "10-region_679 : 4  (P= 0.237972906333401 ) ->  3 (P= 0.358191289505388 )\n",
            "12-region_680 : 2  (P= 0.119916592484809 ) ->  1 (P= 0.56998728355779 )\n",
            "1-region_699 : 4  (P= 0.17678554476983 ) ->  3 (P= 0.351230564409482 )\n",
            "3-region_705 : 5  (P= 0.138031530426655 ) ->  6 (P= 0.29746403147278 )\n",
            "3-region_706 : 4  (P= 0.118594824652238 ) ->  6 (P= 0.234479220714368 )\n",
            "4-region_707 : 2  (P= 0.0588396272818131 ) ->  1 (P= 0.413105217577403 )\n",
            "6-region_710 : 2  (P= 0.211411656214838 ) ->  1 (P= 0.319675212855924 )\n",
            "8-region_714 : 2  (P= 0.176974665424047 ) ->  1 (P= 0.352607227045773 )\n",
            "9-region_716 : 6  (P= 0.292763505998204 ) ->  5 (P= 0.353316490339748 )\n",
            "11-region_717 : 4  (P= 0.235050409376623 ) ->  3 (P= 0.411865912465853 )\n",
            "12-region_721 : 4  (P= 0.11713368710487 ) ->  3 (P= 0.293493580432472 )\n",
            "3-region_726 : 2  (P= 0.126882014260235 ) ->  1 (P= 0.579207586292455 )\n",
            "4-region_727 : 2  (P= 0.137678969025527 ) ->  1 (P= 0.411543819325789 )\n",
            "5-region_728 : 2  (P= 0.342737728746121 ) ->  1 (P= 0.362161500027906 )\n",
            "X-region_732 : 2  (P= 0.216370386058756 ) ->  1 (P= 0.370492649139303 )\n",
            "8-region_733 : 2  (P= 0.185604067842805 ) ->  1 (P= 0.520098874569156 )\n",
            "9-region_734 : 4  (P= 0.165187260493147 ) ->  3 (P= 0.27161729889801 )\n",
            "3-region_741 : 2  (P= 0.191528161051083 ) ->  1 (P= 0.307967302682574 )\n",
            "4-region_742 : 2  (P= 0.191918958570408 ) ->  1 (P= 0.308225612262875 )\n",
            "5-region_743 : 2  (P= 0.153667507904306 ) ->  1 (P= 0.500756705427575 )\n",
            "6-region_744 : 2  (P= 0.116268079864682 ) ->  1 (P= 0.345548266728394 )\n",
            "X-region_747 : 2  (P= 0.192398671279387 ) ->  1 (P= 0.424390370240311 )\n",
            "1-region_753 : 4  (P= 0.260148540297247 ) ->  3 (P= 0.34860284992209 )\n",
            "3-region_756 : 2  (P= 0.218313118852714 ) ->  1 (P= 0.302564613102624 )\n",
            "4-region_757 : 2  (P= 0.131093918012884 ) ->  1 (P= 0.436288801450112 )\n",
            "5-region_758 : 2  (P= 0.218800509433572 ) ->  1 (P= 0.476503451124959 )\n",
            "8-region_763 : 2  (P= 0.2595944867299 ) ->  1 (P= 0.433904566838004 )\n",
            "9-region_764 : 4  (P= 0.0884283183996754 ) ->  3 (P= 0.217811768384223 )\n",
            "12-region_767 : 2  (P= 0.175219747283777 ) ->  1 (P= 0.392338486084562 )\n",
            "4-region_771 : 2  (P= 0.252488283991325 ) ->  1 (P= 0.350087314199273 )\n",
            "X-region_776 : 2  (P= 0.198933602040324 ) ->  1 (P= 0.401314644179244 )\n",
            "8-region_777 : 2  (P= 0.24903585343617 ) ->  1 (P= 0.251576121458134 )\n",
            "12-region_781 : 2  (P= 0.199606365218538 ) ->  4 (P= 0.299934361814559 )\n",
            "1-region_782 : 2  (P= 0.263882293216073 ) ->  3 (P= 0.267918129404798 )\n",
            "1-region_784 : 4  (P= 0.204059459875652 ) ->  3 (P= 0.266543605540319 )\n",
            "2-region_786 : 2  (P= 0.266360288239423 ) ->  3 (P= 0.400961721381145 )\n",
            "3-region_788 : 2  (P= 0.200415649267807 ) ->  1 (P= 0.266817719303142 )\n",
            "5-region_790 : 2  (P= 0.134128486864103 ) ->  1 (P= 0.466560661355394 )\n",
            "7-region_794 : 5  (P= 0.134202834317632 ) ->  4 (P= 0.333716075964186 )\n",
            "X-region_796 : 2  (P= 0.13557111246067 ) ->  1 (P= 0.465415505372764 )\n",
            "9-region_798 : 5  (P= 0.134417935513073 ) ->  4 (P= 0.333492504740297 )\n",
            "10-region_800 : 5  (P= 0.134968390783317 ) ->  6 (P= 0.400969014606499 )\n",
            "1-region_802 : 5  (P= 0.124722688906748 ) ->  6 (P= 0.251659001731851 )\n",
            "2-region_804 : 5  (P= 0.124722688906748 ) ->  6 (P= 0.251659001731851 )\n",
            "2-region_805 : 4  (P= 0.123378844843512 ) ->  3 (P= 0.250991410785825 )\n",
            "4-region_807 : 2  (P= 0.124618328857104 ) ->  3 (P= 0.249374532755791 )\n",
            "X-region_812 : 2  (P= 0.249740426620524 ) ->  1 (P= 0.25015844196856 )\n",
            "9-region_814 : 5  (P= 0.126065700620528 ) ->  3 (P= 0.249701123312902 )\n",
            "10-region_817 : 4  (P= 0.12599856156968 ) ->  5 (P= 0.249035429331423 )\n",
            "1-region_820 : 4  (P= 0.12599856156968 ) ->  5 (P= 0.249035429331423 )\n",
            "5-region_826 : 2  (P= 0.126633087819383 ) ->  3 (P= 0.250906613771838 )\n",
            "10-region_834 : 2  (P= 0.123401691709328 ) ->  1 (P= 0.252241299864004 )\n",
            "12-region_835 : 2  (P= 0.125602969860615 ) ->  1 (P= 0.254405953071525 )\n",
            "5-region_860 : 2  (P= 0.309681730379961 ) ->  3 (P= 0.339964548500964 )\n",
            "11-region_867 : 2  (P= 0.217030729817737 ) ->  3 (P= 0.474451086660427 )\n",
            "4-region_873 : 4  (P= 0.109261185172723 ) ->  1 (P= 0.257444947208122 )\n",
            "6-region_875 : 4  (P= 0.175054872803283 ) ->  3 (P= 0.433807182370605 )\n",
            "9-region_880 : 2  (P= 0.134946009083334 ) ->  1 (P= 0.40588030335842 )\n",
            "10-region_882 : 2  (P= 0.224977823043163 ) ->  3 (P= 0.382000579832458 )\n",
            "4-region_889 : 4  (P= 0.147036762919147 ) ->  3 (P= 0.271675187995741 )\n",
            "5-region_890 : 2  (P= 0.309400348262216 ) ->  3 (P= 0.327365755675575 )\n",
            "7-region_893 : 2  (P= 0.271663089527514 ) ->  1 (P= 0.511177976428904 )\n",
            "7-region_907 : 4  (P= 0.162726127345953 ) ->  3 (P= 0.398697348064127 )\n",
            "8-region_910 : 2  (P= 0.159680115757848 ) ->  1 (P= 0.318866400686507 )\n",
            "12-region_914 : 2  (P= 0.193029937276673 ) ->  1 (P= 0.299144618772521 )\n",
            "3-region_917 : 4  (P= 0.104245675753826 ) ->  3 (P= 0.230293511051304 )\n",
            "4-region_918 : 4  (P= 0.167862048985353 ) ->  3 (P= 0.290572022987121 )\n",
            "11-region_926 : 4  (P= 0.29701418479007 ) ->  3 (P= 0.347880856858127 )\n",
            "1-region_931 : 2  (P= 0.239088839389505 ) ->  3 (P= 0.332307117798466 )\n",
            "4-region_934 : 4  (P= 0.174213692432266 ) ->  3 (P= 0.240202609866855 )\n",
            "11-region_943 : 2  (P= 0.32679944642652 ) ->  3 (P= 0.412906913971689 )\n",
            "12-region_945 : 2  (P= 0.215514753488683 ) ->  1 (P= 0.435117539072345 )\n",
            "2-region_947 : 4  (P= 0.22101357729621 ) ->  5 (P= 0.255543944897444 )\n",
            "9-region_956 : 4  (P= 0.151046190814726 ) ->  5 (P= 0.231373109443362 )\n",
            "10-region_958 : 2  (P= 0.171465732290167 ) ->  1 (P= 0.345645816206934 )\n",
            "12-region_959 : 4  (P= 0.199542757723098 ) ->  3 (P= 0.372438189256246 )\n",
            "1-region_961 : 4  (P= 0.258688977658028 ) ->  3 (P= 0.298452664374789 )\n",
            "4-region_964 : 4  (P= 0.148724085456456 ) ->  3 (P= 0.187101625118904 )\n",
            "5-region_965 : 2  (P= 0.148341167840432 ) ->  3 (P= 0.36924729149797 )\n",
            "8-region_970 : 2  (P= 0.221570064176914 ) ->  1 (P= 0.335520710114798 )\n",
            "1-region_975 : 2  (P= 0.189136652540577 ) ->  3 (P= 0.312135684084975 )\n",
            "1-region_977 : 4  (P= 0.123660580456117 ) ->  3 (P= 0.309839992685504 )\n",
            "4-region_980 : 4  (P= 0.126084367670378 ) ->  5 (P= 0.312290019631579 )\n",
            "7-region_984 : 2  (P= 0.124832461614955 ) ->  3 (P= 0.37582515009013 )\n",
            "8-region_986 : 2  (P= 0.187420938638162 ) ->  3 (P= 0.31307144742556 )\n",
            "10-region_989 : 2  (P= 0.18654167329634 ) ->  1 (P= 0.250815515000267 )\n",
            "12-region_990 : 2  (P= 0.188270114468789 ) ->  1 (P= 0.37533580338121 )\n",
            "1-region_991 : 4  (P= 0.182677776109494 ) ->  5 (P= 0.364418673450332 )\n",
            "2-region_993 : 4  (P= 0.182677776109494 ) ->  5 (P= 0.364418673450332 )\n",
            "8-region_1001 : 4  (P= 0.180909686202978 ) ->  5 (P= 0.182796742524909 )\n",
            "9-region_1002 : 2  (P= 0.0921417054776341 ) ->  1 (P= 0.362408434671359 )\n",
            "1-region_1006 : 4  (P= 0.112633720578415 ) ->  2 (P= 0.22121598939805 )\n",
            "4-region_1012 : 4  (P= 0.111985951952748 ) ->  1 (P= 0.224595258886438 )\n",
            "7-region_1017 : 4  (P= 0.111219600970315 ) ->  3 (P= 0.33352101059958 )\n",
            "9-region_1020 : 5  (P= 0.108735965249315 ) ->  2 (P= 0.224254217125708 )\n",
            "10-region_1023 : 2  (P= 0.11175577544487 ) ->  3 (P= 0.33261736016046 )\n",
            "12-region_1025 : 2  (P= 0.11175577544487 ) ->  3 (P= 0.33261736016046 )\n",
            "3-region_1029 : 2  (P= 0.248377414150658 ) ->  1 (P= 0.251782596723111 )\n",
            "4-region_1030 : 1  (P= 0.247411568413183 ) ->  3 (P= 0.250916130988864 )\n",
            "6-region_1032 : 4  (P= 0.12483477710735 ) ->  3 (P= 0.251913546884676 )\n",
            "7-region_1034 : 6  (P= 0.249564424605358 ) ->  5 (P= 0.250144084132924 )\n",
            "10-region_1082 : 5  (P= 0.144875548770908 ) ->  6 (P= 0.282206028356585 )\n",
            "8-region_1095 : 4  (P= 0.189303864179653 ) ->  3 (P= 0.248573251786051 )\n",
            "6-region_1105 : 2  (P= 0.161922030684451 ) ->  3 (P= 0.449599712602403 )\n",
            "10-region_1112 : 4  (P= 0.147447750190247 ) ->  3 (P= 0.386952784728723 )\n",
            "7-region_1121 : 2  (P= 0.276511202407559 ) ->  3 (P= 0.340712962168781 )\n",
            "X-region_1122 : 4  (P= 0.215877718774882 ) ->  3 (P= 0.236387866753323 )\n",
            "9-region_1124 : 5  (P= 0.314435275540801 ) ->  6 (P= 0.318096713876685 )\n",
            "10-region_1126 : 4  (P= 0.282578247634745 ) ->  3 (P= 0.375569420337284 )\n",
            "4-region_1131 : 2  (P= 0.183729731971913 ) ->  1 (P= 0.334184840982857 )\n",
            "5-region_1147 : 2  (P= 0.316474528379823 ) ->  3 (P= 0.316680445313208 )\n",
            "9-region_1153 : 4  (P= 0.205968113707141 ) ->  3 (P= 0.261943243007867 )\n",
            "10-region_1155 : 4  (P= 0.177445747575994 ) ->  6 (P= 0.254578968694951 )\n",
            "12-region_1172 : 2  (P= 0.157192243294034 ) ->  1 (P= 0.276837457637366 )\n",
            "9-region_1183 : 4  (P= 0.158602183807343 ) ->  3 (P= 0.250216246144091 )\n",
            "9-region_1197 : 2  (P= 0.20247464007232 ) ->  3 (P= 0.321088225469875 )\n",
            "1-region_1201 : 2  (P= 0.281382277661728 ) ->  1 (P= 0.441537931642319 )\n",
            "8-region_1210 : 4  (P= 0.196687438623401 ) ->  3 (P= 0.313534225908177 )\n",
            "12-region_1214 : 4  (P= 0.192410473303185 ) ->  3 (P= 0.438754294982852 )\n",
            "8-region_1227 : 4  (P= 0.202434462736568 ) ->  3 (P= 0.257795526754182 )\n",
            "2-region_1248 : 2  (P= 0.28487414755127 ) ->  3 (P= 0.396169308792251 )\n",
            "3-region_1249 : 2  (P= 0.165785068485549 ) ->  1 (P= 0.313240208868005 )\n",
            "4-region_1250 : 2  (P= 0.144419305808824 ) ->  1 (P= 0.419653350060016 )\n",
            "5-region_1252 : 2  (P= 0.169878304258809 ) ->  3 (P= 0.371646600734274 )\n",
            "7-region_1256 : 4  (P= 0.167315372700467 ) ->  3 (P= 0.395271329583443 )\n",
            "9-region_1259 : 4  (P= 0.104285868325543 ) ->  3 (P= 0.353880604589929 )\n",
            "10-region_1261 : 5  (P= 0.167917294833744 ) ->  3 (P= 0.293143227413872 )\n",
            "3-region_1265 : 4  (P= 0.163164771225264 ) ->  5 (P= 0.269663822662441 )\n",
            "5-region_1267 : 4  (P= 0.207203804312266 ) ->  5 (P= 0.227298193583215 )\n",
            "8-region_1272 : 5  (P= 0.229088531586279 ) ->  6 (P= 0.25058764901388 )\n",
            "4-region_1280 : 2  (P= 0.170277600590281 ) ->  1 (P= 0.326066769970673 )\n",
            "12-region_1290 : 2  (P= 0.198902524234445 ) ->  1 (P= 0.449521799363502 )\n",
            "4-region_1295 : 2  (P= 0.227953938411826 ) ->  3 (P= 0.342002035424784 )\n",
            "1-region_1307 : 2  (P= 0.373136500821113 ) ->  1 (P= 0.376834562240272 )\n",
            "2-region_1308 : 2  (P= 0.186380186823256 ) ->  1 (P= 0.251400095804457 )\n",
            "10-region_1319 : 4  (P= 0.187041610246259 ) ->  3 (P= 0.377928294805532 )\n",
            "1-region_1321 : 2  (P= 0.197928047559057 ) ->  1 (P= 0.300416987544011 )\n",
            "2-region_1323 : 2  (P= 0.100551193283198 ) ->  1 (P= 0.299320443321652 )\n",
            "X-region_1330 : 4  (P= 0.199354701557877 ) ->  6 (P= 0.202133911047324 )\n",
            "8-region_1331 : 5  (P= 0.201839844615743 ) ->  3 (P= 0.298581249464737 )\n",
            "11-region_1334 : 4  (P= 0.29777194088908 ) ->  3 (P= 0.299810494639734 )\n",
            "10-region_1335 : 4  (P= 0.198603506718463 ) ->  2 (P= 0.302085904253615 )\n",
            "12-region_1337 : 5  (P= 0.100169660377912 ) ->  6 (P= 0.301460549144924 )\n",
            "1-region_1338 : 2  (P= 0.197565824047885 ) ->  1 (P= 0.301012128946364 )\n",
            "2-region_1342 : 4  (P= 0.0984789853485853 ) ->  3 (P= 0.402223178962523 )\n",
            "6-region_1346 : 1  (P= 0.298853215752141 ) ->  3 (P= 0.299566491194817 )\n",
            "1-region_1355 : 2  (P= 0.110937692894237 ) ->  3 (P= 0.335903474177799 )\n",
            "5-region_1361 : 4  (P= 0.222034821509058 ) ->  3 (P= 0.332747875810516 )\n",
            "7-region_1365 : 4  (P= 0.218340881069897 ) ->  5 (P= 0.223490157046081 )\n",
            "X-region_1366 : 5  (P= 0.222742361545082 ) ->  6 (P= 0.223816697872273 )\n",
            "9-region_1368 : 2  (P= 0.221137639106631 ) ->  1 (P= 0.334540586211101 )\n",
            "11-region_1370 : 2  (P= 0.223486751917125 ) ->  1 (P= 0.224441539318089 )\n",
            "1-region_1378 : 4  (P= 0.122942187093408 ) ->  3 (P= 0.252751388727581 )\n",
            "3-region_1381 : 2  (P= 0.124659457613695 ) ->  3 (P= 0.25232599730902 )\n",
            "6-region_1384 : 5  (P= 0.126815746481796 ) ->  3 (P= 0.252314414706138 )\n",
            "8-region_1388 : 1  (P= 0.248439686082223 ) ->  2 (P= 0.253719939221304 )\n",
            "9-region_1389 : 5  (P= 0.12527900142488 ) ->  6 (P= 0.252034050904212 )\n",
            "12-region_1412 : 5  (P= 0.144517489251129 ) ->  6 (P= 0.285918652968831 )\n",
            "INFO [2025-05-07 17:41:32] Creating Plots for CNV and cell Probabilities.\n",
            "INFO [2025-05-07 17:46:10] ::plot_cnv:Start\n",
            "INFO [2025-05-07 17:46:10] ::plot_cnv:Current data dimensions (r,c)=477,2730 Total=511631.072568956 Min=0 Max=0.99138363885186.\n",
            "INFO [2025-05-07 17:46:10] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
            "INFO [2025-05-07 17:46:10] plot_cnv_observation:Start\n",
            "INFO [2025-05-07 17:46:10] Observation data size: Cells= 1367 Genes= 477\n",
            "INFO [2025-05-07 17:46:10] plot_cnv_observation:Writing observation groupings/color.\n",
            "INFO [2025-05-07 17:46:11] plot_cnv_observation:Done writing observation groupings/color.\n",
            "INFO [2025-05-07 17:46:11] plot_cnv_observation:Writing observation heatmap thresholds.\n",
            "INFO [2025-05-07 17:46:11] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
            "INFO [2025-05-07 17:46:11] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-05-07 17:46:11] Quantiles of plotted data range: 0,0,0.769692534369413,0.922996952074611,0.9906421637069\n",
            "INFO [2025-05-07 17:46:11] plot_cnv_references:Start\n",
            "INFO [2025-05-07 17:46:11] Reference data size: Cells= 1363 Genes= 477\n",
            "INFO [2025-05-07 17:46:11] plot_cnv_references:Number reference groups= 2\n",
            "INFO [2025-05-07 17:46:11] plot_cnv_references:Plotting heatmap.\n",
            "INFO [2025-05-07 17:46:11] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-05-07 17:46:11] Quantiles of plotted data range: 0,0,0,0.659287037831219,0.99138363885186\n",
            "INFO [2025-05-07 17:46:14] ::plot_cnv:Start\n",
            "INFO [2025-05-07 17:46:14] ::plot_cnv:Current data dimensions (r,c)=477,2730 Total=3987735 Min=1 Max=6.\n",
            "INFO [2025-05-07 17:46:14] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
            "INFO [2025-05-07 17:46:14] plot_cnv_observation:Start\n",
            "INFO [2025-05-07 17:46:14] Observation data size: Cells= 1367 Genes= 477\n",
            "INFO [2025-05-07 17:46:14] plot_cnv_observation:Writing observation groupings/color.\n",
            "INFO [2025-05-07 17:46:14] plot_cnv_observation:Done writing observation groupings/color.\n",
            "INFO [2025-05-07 17:46:14] plot_cnv_observation:Writing observation heatmap thresholds.\n",
            "INFO [2025-05-07 17:46:14] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
            "INFO [2025-05-07 17:46:14] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-05-07 17:46:14] Quantiles of plotted data range: 1,2,3,4,6\n",
            "INFO [2025-05-07 17:46:15] plot_cnv_references:Start\n",
            "INFO [2025-05-07 17:46:15] Reference data size: Cells= 1363 Genes= 477\n",
            "INFO [2025-05-07 17:46:15] plot_cnv_references:Number reference groups= 2\n",
            "INFO [2025-05-07 17:46:15] plot_cnv_references:Plotting heatmap.\n",
            "INFO [2025-05-07 17:46:15] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-05-07 17:46:15] Quantiles of plotted data range: 1,3,3,3,6\n",
            "INFO [2025-05-07 17:46:16] \n",
            "\n",
            "\tSTEP 20: Converting HMM-based CNV states to repr expr vals\n",
            "\n",
            "INFO [2025-05-07 17:46:18] ::plot_cnv:Start\n",
            "INFO [2025-05-07 17:46:18] ::plot_cnv:Current data dimensions (r,c)=477,2730 Total=1382987.5 Min=0 Max=3.\n",
            "INFO [2025-05-07 17:46:18] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
            "INFO [2025-05-07 17:46:18] plot_cnv_observation:Start\n",
            "INFO [2025-05-07 17:46:18] Observation data size: Cells= 1367 Genes= 477\n",
            "INFO [2025-05-07 17:46:19] plot_cnv_observation:Writing observation groupings/color.\n",
            "INFO [2025-05-07 17:46:19] plot_cnv_observation:Done writing observation groupings/color.\n",
            "INFO [2025-05-07 17:46:19] plot_cnv_observation:Writing observation heatmap thresholds.\n",
            "INFO [2025-05-07 17:46:19] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
            "INFO [2025-05-07 17:46:19] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-05-07 17:46:19] Quantiles of plotted data range: 0,0.5,1,1.5,3\n",
            "INFO [2025-05-07 17:46:19] plot_cnv_references:Start\n",
            "INFO [2025-05-07 17:46:19] Reference data size: Cells= 1363 Genes= 477\n",
            "INFO [2025-05-07 17:46:19] plot_cnv_references:Number reference groups= 2\n",
            "INFO [2025-05-07 17:46:19] plot_cnv_references:Plotting heatmap.\n",
            "INFO [2025-05-07 17:46:19] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-05-07 17:46:19] Quantiles of plotted data range: 0,1,1,1,3\n",
            "INFO [2025-05-07 17:46:20] \n",
            "\n",
            "\tSTEP 22: Denoising\n",
            "\n",
            "INFO [2025-05-07 17:46:20] ::process_data:Remove noise, noise threshold defined via ref mean sd_amplifier:  1.5\n",
            "INFO [2025-05-07 17:46:20] denoising using mean(normal) +- sd_amplifier * sd(normal) per gene per cell across all data\n",
            "INFO [2025-05-07 17:46:20] :: **** clear_noise_via_ref_quantiles **** : removing noise between bounds:  0.881298801280801 - 1.12639483334381\n",
            "INFO [2025-05-07 17:46:24] \n",
            "\n",
            "## Making the final infercnv heatmap ##\n",
            "INFO [2025-05-07 17:46:25] ::plot_cnv:Start\n",
            "INFO [2025-05-07 17:46:25] ::plot_cnv:Current data dimensions (r,c)=477,2730 Total=1315924.74136779 Min=0.311386364997398 Max=9.12526404188546.\n",
            "INFO [2025-05-07 17:46:25] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
            "INFO [2025-05-07 17:46:25] plot_cnv(): auto thresholding at: (0.714485 , 1.285515)\n",
            "INFO [2025-05-07 17:46:25] plot_cnv_observation:Start\n",
            "INFO [2025-05-07 17:46:25] Observation data size: Cells= 1367 Genes= 477\n",
            "INFO [2025-05-07 17:46:25] plot_cnv_observation:Writing observation groupings/color.\n",
            "INFO [2025-05-07 17:46:25] plot_cnv_observation:Done writing observation groupings/color.\n",
            "INFO [2025-05-07 17:46:25] plot_cnv_observation:Writing observation heatmap thresholds.\n",
            "INFO [2025-05-07 17:46:25] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
            "INFO [2025-05-07 17:46:25] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-05-07 17:46:25] Quantiles of plotted data range: 0.714485226991113,1.0038468173123,1.0038468173123,1.0038468173123,1.28551477300889\n",
            "INFO [2025-05-07 17:46:25] plot_cnv_references:Start\n",
            "INFO [2025-05-07 17:46:25] Reference data size: Cells= 1363 Genes= 477\n",
            "INFO [2025-05-07 17:46:26] plot_cnv_references:Number reference groups= 2\n",
            "INFO [2025-05-07 17:46:26] plot_cnv_references:Plotting heatmap.\n",
            "INFO [2025-05-07 17:46:26] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
            "INFO [2025-05-07 17:46:26] Quantiles of plotted data range: 0.714485226991113,1.0038468173123,1.0038468173123,1.0038468173123,1.28551477300889\n"
          ]
        }
      ]
    }
  ]
}
